---
title: "Check RNAseq data"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Check the bulk RNAseq data from May release.

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(readxl)

theme_set(theme_classic())

path = "~/research/data/MorPhiC/May-2024/JAX_RNAseq2_ExtraEmbryonic/"
dat_path = paste0(path, "processed/release98-gencode32/Tables")

read_header <- function(file_name, sheet){
  
  meta_header = read_excel(file_name, sheet = sheet, 
                         skip = 3, n_max = 1, col_names = FALSE)
  meta_header = as.character(meta_header)
  meta_header = gsub("cell_line.", "", meta_header, fixed = TRUE)
  meta_header = gsub("differentiated_cell_line.", "", meta_header, fixed = TRUE)
  meta_header = gsub("biomaterial_core.", "", meta_header, fixed = TRUE)
  meta_header = gsub("gene_expression_alteration_protocol.protocol_core.", 
                     "", meta_header, fixed = TRUE)
  meta_header = gsub(".text", "", meta_header, fixed = TRUE)
  
  meta_header = gsub("gene_expression_alteration_protocol.", "", 
                     meta_header, fixed = TRUE)

  meta_header
}

```

## Read in meta data

```{r fig.width = 3, fig.height=3}
meta_file = "data/JAX_RNAseq2_ExtraEmbryonic_metadata.xlsx"

cl_header = read_header(meta_file, sheet = "Cell line ")
cl_header

cl = read_excel(meta_file, sheet = "Cell line ", skip = 5, col_names = FALSE)
names(cl) = cl_header
cl = as.data.frame(cl)
dim(cl)
cl[1:2,]

dc_header = read_header(meta_file, sheet = "Differentiated cell line")
dc_header

dc = read_excel(meta_file, sheet = "Differentiated cell line", 
                skip = 5, col_names = FALSE)
names(dc) = dc_header
dc = as.data.frame(dc)
dim(dc)
dc[1:2,]

## remove columns 3:8, differentiated_ncbi_taxon_id to 
## differentiated_biosamples_accession
dc = dc[,-(3:8)]
dim(dc)
dc[1:2,]

eas_header = read_header(meta_file, sheet = "Expression alteration strategy")
eas_header
eas_header = c(eas_header, "altered_locus", "guide_sequence")

eas = read_excel(meta_file, sheet = "Expression alteration strategy", 
                skip = 5, col_names = FALSE)
names(eas) = eas_header
eas = as.data.frame(eas)
dim(eas)
eas[1:2,]
```

## Check cell lines

```{r}
any(duplicated(cl$biomaterial_id))
table(cl$protocol_id)
table(cl$type)
table(cl$derived_cell_line_accession)
table(cl$zygosity)
sort(table(cl$clone_id), decreasing = TRUE)

table(dc$biomaterial_id %in% cl$biomaterial_id)
table(cl$biomaterial_id %in% dc$biomaterial_id)
setdiff(dc$biomaterial_id, cl$biomaterial_id)
setdiff(cl$biomaterial_id, dc$biomaterial_id)

table(table(dc$biomaterial_id))
table(dc$differentiated_timepoint_value)
table(dc$differentiated_terminally_differentiated)
table(dc$differentiation_protocol.protocol_core.protocol_id, 
      dc$model_organ)
```

## Check Expression alteration strategy
```{r}
table(eas$protocol_id)
table(eas$allele_specific)
table(eas$protocol_id, eas$altered_gene_symbols)
table(eas$protocol_id, eas$crispr.sgrna_target)
```

## Extract KO strategy information

```{r}
dc[["ko_strategy"]] = 
  str_extract(dc$differentiated_biomaterial_description, 
              "\\(KO\\)|\\(CE\\)|\\(PTC\\)")
dc[["ko_strategy"]] = gsub("\\(|\\)", "", dc[["ko_strategy"]])

dc[["ko_gene"]] = str_extract(dc$differentiated_biomaterial_description, 
                                  "(?<=KOLF2.2 deleted ).\\S+")
table(dc$ko_strategy, dc$ko_gene, useNA = 'ifany')
```


## Read in library preparation information
```{r}
lib_header = read_header(meta_file, sheet = "Library preparation")
lib_header
lib_header = c(lib_header, 
               c("fragment_size", "input_amount", "input_unit", 
                 "final_yield", "final_yield_unit", 
                 "lib_concentration", "lib_concentration_unit",
                 "PCR_cycle", "PCR_cycle_sample_index"))

lib = read_excel(meta_file, sheet = "Library preparation", 
                skip = 5, col_names = FALSE)
dim(lib)
names(lib) = lib_header
lib = as.data.frame(lib)
dim(lib)
lib[1:2,]

seq_header = read_header(meta_file, sheet = "Sequence file")
seq_header

seq = read_excel(meta_file, sheet = "Sequence file", 
                skip = 5, col_names = FALSE)
names(seq) = seq_header
seq = as.data.frame(seq)
seq$file_id = gsub("_R(1|2)_001.fastq.gz", "", 
                   seq$sequence_file.file_core.file_name)
dim(seq)
seq[1:2,]

seq = unique(seq[,c("file_id", "library_preparation.biomaterial_id", 
                    "sequencing_protocol.protocol_core.protocol_id", 
                    "sequence_file.run_id")])
dim(seq)
seq[1:2,]

table(seq$sequence_file.run_id)

intersect(names(dc), names(lib))
meta = merge(dc, lib)
dim(meta)
meta[1:2,]

meta = merge(meta, seq)
dim(meta)
meta[1:2,]

update_name <- function(df1, old_name, new_name){
  ww1 = which(names(df1) == old_name)
  stopifnot(length(ww1) == 1)
  names(df1)[ww1] = new_name
  df1
}

meta = update_name(meta, "library_preparation.biomaterial_id", 
                   "lib_biomaterial_id")

meta = update_name(meta, "differentiation_protocol.protocol_core.protocol_id", 
                   "differentiation_protocol")

meta = update_name(meta, "library_preparation_protocol.protocol_core.protocol_id", 
                   "library_preparation_protocol")

meta = update_name(meta, "dissociation_protocol.protocol_core.protocol_id", 
                   "dissociation_protocol")

meta = update_name(meta, "sequencing_protocol.protocol_core.protocol_id", 
                   "sequencing_protocol")

meta = update_name(meta, "sequence_file.run_id", "run_id")

intersect(names(meta), names(cl))

meta = merge(meta, cl)
dim(meta)
meta[1:2,]

table(meta$ko_strategy, useNA = "ifany")
meta$ko_strategy[which(is.na(meta$ko_strategy))] = "WT"

table(meta$ko_gene, useNA = "ifany")
meta$ko_gene[which(is.na(meta$ko_gene))] = "WT"

fwrite(meta, file = "data/JAX_RNAseq2_ExtraEmbryonic_meta_data.tsv", sep="\t")
```

## Check count data

### genesCounts.csv: count data have some extra columns for ISL1

```{r fig.width = 5.5, fig.height=4}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

setdiff(names(cts)[-1], meta$file_id)

dat_path_pre = "~/research/data/MorPhiC/JAX_RNAseq_ExtraEmbryonic/processed/Tables"

cts_pre = fread(file.path(dat_path_pre, "genesCounts.csv"), data.table = FALSE)
dim(cts_pre)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

sample_names = intersect(names(cts), names(cts_pre))
sample_names

gene_names = intersect(cts$Name, cts_pre$Name)
length(gene_names)

cts_repeat = cts[match(gene_names, cts$Name), sample_names]
cts_pre_repeat = cts_pre[match(gene_names, cts_pre$Name), sample_names]

cts_repeat[1:10,1:2]
cts_pre_repeat[1:10,1:2]

cts_repeat[1:10,c(1,4)]
cts_pre_repeat[1:10,c(1,4)]

df1 = data.frame(ISL1_PTC_pre = cts_pre_repeat[,2], 
                 ISL1_PTC_may = cts_repeat[,2])

ggplot(data = df1, mapping = aes(x = log10(ISL1_PTC_pre + 1), 
                                 y = log10(ISL1_PTC_may + 1))) +
  geom_pointdensity() + scale_color_viridis() + 
  ggtitle("genesCounts") + 
  geom_abline(slope = 1, intercept = 0, col = "orange")

```

### featureCounts.csv: count data have some extra columns for ISL1

```{r fig.width = 5.5, fig.height=4}
cts = fread(file.path(dat_path, "featureCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

setdiff(names(cts)[-1], meta$file_id)
cts[1:10,c("name", "ISL1_CE_B05_GT23-12151_CACGAGCG-TGGCAGAT_S27_L001")]

dat_path_pre = "~/research/data/MorPhiC/JAX_RNAseq_ExtraEmbryonic/processed/Tables"

cts_pre = fread(file.path(dat_path_pre, "featureCounts.csv"), data.table = FALSE)
dim(cts_pre)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

sample_names = intersect(names(cts), names(cts_pre))
sample_names

gene_names = intersect(cts$name, cts_pre$name)
length(gene_names)

cts_repeat = cts[match(gene_names, cts$name), sample_names]
cts_pre_repeat = cts_pre[match(gene_names, cts_pre$name), sample_names]

cts_repeat[1:10,1:2]
cts_pre_repeat[1:10,1:2]

cts_repeat[1:10,c(1,4)]
cts_pre_repeat[1:10,c(1,4)]

df1 = data.frame(ISL1_PTC_pre = cts_pre_repeat[,2], 
                 ISL1_PTC_may = cts_repeat[,2])

ggplot(data = df1, mapping = aes(x = log10(ISL1_PTC_pre + 1), 
                                 y = log10(ISL1_PTC_may + 1))) +
  geom_pointdensity() + scale_color_viridis() + 
  ggtitle("featureCounts") + 
  geom_abline(slope = 1, intercept = 0, col = "orange")
```

```{r}
gc()
sessionInfo()
```



