---
title: "Check RNAseq data"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(readxl)

theme_set(theme_classic())

dat_path = "~/research/data/MorPhiC/December-2023/JAX_RNAseq_ExtraEmbryonic/"
dat_path = paste0(dat_path, "processed/Tables")

read_header <- function(file_name, sheet){
  
  meta_header = read_excel(file_name, sheet = sheet, 
                         skip = 3, n_max = 1, col_names = FALSE)
  meta_header = as.character(meta_header)
  meta_header = gsub("cell_line.", "", meta_header, fixed = TRUE)
  meta_header = gsub("differentiated_cell_line.", "", meta_header, fixed = TRUE)
  meta_header = gsub("biomaterial_core.", "", meta_header, fixed = TRUE)
  meta_header = gsub("gene_expression_alteration_protocol.protocol_core.", 
                     "", meta_header, fixed = TRUE)
  meta_header = gsub(".text", "", meta_header, fixed = TRUE)
  
  meta_header = gsub("gene_expression_alteration_protocol.", "", 
                     meta_header, fixed = TRUE)

  meta_header
}

```

## Read in meta data

```{r fig.width = 3, fig.height=3}
cl_header = read_header("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                         sheet = "Cell line ")
cl_header

cl = read_excel("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                sheet = "Cell line ", 
                skip = 5, col_names = FALSE)
names(cl) = cl_header
cl = as.data.frame(cl)
dim(cl)
cl[1:2,]

dc_header = read_header("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                         sheet = "Differentiated cell line")
dc_header

dc = read_excel("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                sheet = "Differentiated cell line", 
                skip = 5, col_names = FALSE)
names(dc) = dc_header
dc = as.data.frame(dc)
dim(dc)
dc[1:2,]

eas_header = read_header("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                         sheet = "Expression alteration strategy")
eas_header

eas = read_excel("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                sheet = "Expression alteration strategy", 
                skip = 5, col_names = FALSE)
names(eas) = eas_header
eas = as.data.frame(eas)
dim(eas)
eas[1:2,]
```

## Check cell lines

```{r}
any(duplicated(cl$biomaterial_id))
table(cl$protocol_id)
table(cl$type)
table(cl$derived_cell_line_accession)
table(cl$zygosity)
sort(table(cl$clone_id), decreasing = TRUE)

table(dc$biomaterial_id %in% cl$biomaterial_id)
table(cl$biomaterial_id %in% dc$biomaterial_id)
setdiff(dc$biomaterial_id, cl$biomaterial_id)
setdiff(cl$biomaterial_id, dc$biomaterial_id)

table(table(dc$biomaterial_id))
table(dc$differentiated_timepoint_value)
table(dc$differentiated_terminally_differentiated)
table(dc$differentiation_protocol.protocol_core.protocol_id, 
      dc$model_organ)
```

## Check Expression alteration strategy
```{r}
table(eas$protocol_id)
table(eas$allele_specific)
table(eas$protocol_id, eas$altered_gene_symbols)
table(eas$protocol_id, eas$crispr.sgrna_target)
```

## Extract KO strategy information

```{r}
dc[["ko_strategy"]] = 
  str_extract(dc$differentiated_biomaterial_description, 
              "\\(KO\\)|\\(CE\\)|\\(PTC\\)")
dc[["ko_strategy"]] = gsub("\\(|\\)", "", dc[["ko_strategy"]])

dc[["ko_gene"]] = str_extract(dc$differentiated_biomaterial_description, 
                                  "(?<=KOLF2.2 deleted ).\\S+")
table(dc$ko_strategy, dc$ko_gene, useNA = 'ifany')
```


## Read in library preparation information
```{r}
lib_header = read_header("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                         sheet = "Library preparation")
lib_header

lib = read_excel("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                sheet = "Library preparation", 
                skip = 5, col_names = FALSE)
names(lib) = lib_header
lib = as.data.frame(lib)
dim(lib)
lib[1:2,]

seq_header = read_header("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                         sheet = "Sequence file")
seq_header

seq = read_excel("data/JAX_RNAseq_ExtraEmbryonic_metadata.xlsx", 
                sheet = "Sequence file", 
                skip = 5, col_names = FALSE)
names(seq) = seq_header
seq = as.data.frame(seq)
seq$file_id = gsub("_R(1|2)_001.fastq.gz", "", 
                   seq$sequence_file.file_core.file_name)
dim(seq)
seq[1:2,]

seq = unique(seq[,c("file_id", "library_preparation.biomaterial_id")])
dim(seq)
seq[1:2,]

meta = merge(dc, lib[,c("library_preparation.biomaterial_id", 
                      "differentiated_biomaterial_id")])
dim(meta)
meta[1:2,]

meta = merge(meta, seq)
dim(meta)
meta[1:2,]

table(meta$ko_strategy, useNA = "ifany")
meta$ko_strategy[which(is.na(meta$ko_strategy))] = "WT"

table(meta$ko_gene, useNA = "ifany")
meta$ko_gene[which(is.na(meta$ko_gene))] = "WT"

fwrite(meta, file = "data/JAX_RNAseq_ExtraEmbryonic_meta.tsv", sep="\t")
```

# Compare three versions of gene expression data

## Read in feature count data

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "featureCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

setequal(names(cts)[-1], meta$file_id)

meta = meta[match(names(cts)[-1], meta$file_id),]
stopifnot(all(names(cts)[-1] == meta$file_id))

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$name
meta$`feature_cts_rd`  = colSums(cts_dat)/1e6
meta$`feature_cts_q75` = apply(cts_dat, 2, quantile, probs = 0.75)

ggplot(meta, aes(x=feature_cts_rd)) + ggtitle("Feature counts") + 
    geom_histogram(fill="lightblue", color="darkblue", bins=15) + 
  xlab("read depth (million) from featureCounts")

ggplot(meta, aes(x=feature_cts_rd, y=feature_cts_q75, 
                 color=ko_strategy)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Feature counts") + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "KO type") + scale_color_brewer(palette = "Set1")
```

## Read in `genesCounts` data

```{r fig.width = 4, fig.height=3}
geneCts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(geneCts)
geneCts[1:2, c(1:2, (ncol(geneCts)-1):ncol(geneCts))]

table(geneCts$Name == cts$name)
setequal(geneCts$Name, cts$name)

geneCts = geneCts[match(cts$name, geneCts$Name),]
stopifnot(all(geneCts$Name == cts$name))

stopifnot(all(names(geneCts)[-1] == names(cts)[-1]))

geneCts_dat = data.matrix(geneCts[,-1])
rownames(geneCts_dat) = cts$name

meta$`genes_cts_rd`  = colSums(geneCts_dat)/1e6
meta$`genes_cts_q75` = apply(geneCts_dat, 2, quantile, probs = 0.75)

ggplot(meta, aes(x=genes_cts_rd, y=genes_cts_q75, 
                 color=ko_strategy)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "KO type") + scale_color_brewer(palette = "Set1")

```

### Compare read-depth

```{r fig.width = 3, fig.height=3}
ggplot(meta, aes(x=feature_cts_rd, y=genes_cts_rd)) + 
  geom_abline(slope = 1, intercept = 0, color = "orange", linewidth=1.5) + 
  geom_point(size = 1.5, alpha = 0.6) + ggtitle("read depth (million)") + 
  xlab("feature counts") + ylab("gene counts")

ggplot(meta, aes(x=feature_cts_q75, y=genes_cts_q75)) + 
  geom_abline(slope = 1, intercept = 0, color = "orange", linewidth=1.5) + 
  geom_point(size = 1.5, alpha = 0.6) + ggtitle("75 percentile of counts") + 
  xlab("feature counts") + ylab("gene counts")
```

## Read in `genesRawCounts` data
```{r fig.width=4, fig.height=3}
geneRawCts = fread(file.path(dat_path, "genesRawCounts.csv"), 
                   data.table = FALSE)
dim(geneRawCts)
geneRawCts[1:2, c(1:2, (ncol(geneRawCts)-1):ncol(geneRawCts))]

table(geneRawCts$Name == cts$name)
setequal(geneRawCts$Name, cts$name)

geneRawCts = geneRawCts[match(cts$name, geneRawCts$Name),]
stopifnot(all(geneRawCts$Name == cts$name))

stopifnot(all(names(geneRawCts)[-1] == names(cts)[-1]))

geneRawCts_dat = data.matrix(geneRawCts[,-1])
rownames(geneRawCts_dat) = cts$name

meta$`genes_raw_cts_rd`  = colSums(geneRawCts_dat)/1e6
meta$`genes_raw_cts_q75` = apply(geneRawCts_dat, 2, quantile, probs = 0.75)

ggplot(meta, aes(x=genes_raw_cts_rd, y=genes_raw_cts_q75, 
                 color=ko_strategy)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene raw counts") + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "KO type") + scale_color_brewer(palette = "Set1")
```

### Compare read-depth

```{r fig.width=3, fig.height=3}
ggplot(meta, aes(x=genes_cts_rd, y=genes_raw_cts_rd)) + 
  geom_abline(slope = 1, intercept = 0, color = "orange", linewidth=1.5) + 
  geom_point(alpha = 0.6) + ggtitle("read depth (million)") + 
  xlab("genes counts") + ylab("genes raw counts")

ggplot(meta, aes(x=genes_cts_q75, y=genes_raw_cts_q75)) + 
  geom_abline(slope = 1, intercept = 0, color = "orange", linewidth=1.5) + 
  geom_point(alpha = 0.6) + ggtitle("75 percentile of counts") + 
  xlab("genes counts") + ylab("genes raw counts")
```

### Filter genes

Keep genes that are expressed in at least 25% of samples. 

```{r fig.width=4, fig.height=3}
cts_gene_q75 = apply(cts_dat, 1, quantile, probs = 0.75)
geneCts_gene_q75 = apply(geneCts_dat, 1, quantile, probs = 0.75)
geneRawCts_gene_q75 = apply(geneRawCts_dat, 1, quantile, probs = 0.75)

table(cts_gene_q75 > 0, geneCts_gene_q75 > 0)
table(geneRawCts_gene_q75 > 0, geneCts_gene_q75 > 0)

w2kp = which(geneCts_gene_q75 > 0)

cts_dat = cts_dat[w2kp,]
geneCts_dat = geneCts_dat[w2kp,]
geneRawCts_dat = geneRawCts_dat[w2kp,]

dim(cts_dat)
dim(geneCts_dat)
dim(geneRawCts_dat)
```

## Check gene-by-gene expression correlation

### Gene-by-gene correlation between geneCounts and gene raw counts

```{r fig.width = 4, fig.height=3}
cr_matrix = cor(t(geneCts_dat), t(geneRawCts_dat), method = "spearman")
dim(cr_matrix)

cr_diag = diag(cr_matrix)
cr_off_diag = c(cr_matrix[upper.tri(cr_matrix)])

length(cr_diag)
length(cr_off_diag)
set.seed(123)
cr_off_diag = sample(cr_off_diag, 1e6)
summary(cr_diag)
summary(cr_off_diag)

df1 = data.frame(cr = cr_diag)
ggplot(df1, aes(x = cr)) + 
  ggtitle("corr. of geneCounts vs. genesRawCounts") + 
  geom_histogram(breaks = seq(0.94, 1, by=0.001), 
                 fill = "lightblue", color="darkblue", alpha = 0.5) + 
  xlab("Spearman Correlation")

```

### Gene-by-gene correlation between featureCounts and gene raw counts

```{r fig.width = 4, fig.height=3}
cr_matrix = cor(t(cts_dat), t(geneCts_dat), method = "spearman")
dim(cr_matrix)

cr_diag = diag(cr_matrix)
cr_off_diag = c(cr_matrix[upper.tri(cr_matrix)])

length(cr_diag)
length(cr_off_diag)
set.seed(123)
cr_off_diag = sample(cr_off_diag, 1e6)
summary(cr_diag)
summary(cr_off_diag)

df1 = data.frame(cr = cr_diag)
df2 = data.frame(cr_off_diag = cr_off_diag)

ggplot(df1, aes(x = cr)) + 
  ggtitle("corr. of featureCoutns vs. genesCounts") + 
  geom_histogram(aes(y = after_stat(density)), bins = 30, 
                 fill = "lightblue", color="darkblue", alpha = 0.5) + 
  geom_density(data = df2, aes(x = cr_off_diag), color = "red", size = 1) + 
  xlab("Spearman Correlation")

df1$mean = apply(cts_dat, 1, mean)
ggplot(df1, mapping = aes(x = log10(mean), y = cr)) +
  geom_pointdensity() + ylab("Spearman correlation") + 
  scale_color_viridis()

sum(cr_diag < 0.5)
sum(cr_diag < 0.1)
sum(cr_diag < 0 )
```

### illustrate 10 genes with low correlation
```{r fig.width = 4, fig.height=3}
set.seed(123)
w2draw = sample(which(cr_diag < 0.1), size = 10)

for(w1 in w2draw){
  gene = rownames(cts_dat)[w1]
  df1 = data.frame(featureCounts = cts_dat[w1,], 
                   geneConunts = geneCts_dat[w1,])
  g1 = ggplot(df1, aes(x = log10(featureCounts + 1), 
                       y = log10(geneConunts + 1))) + 
    geom_point(alpha = 0.5, size=2) + ggtitle(gene)
  print(g1)
}
```


## PCA

### All samples

```cts_dat_n``` is normalized gene expression data, each row is a gene and each sample is a sample.

```{r fig.width = 7, fig.height=3.5}
cts_dat_n = t(geneCts_dat)
cts_dat_n = log(median(meta$feature_cts_q75)*cts_dat_n/meta$feature_cts_q75 + 1)

dim(cts_dat_n)
sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2

eigen_df = data.frame(PC = 1:length(eigen_vals), Eigenvalue = eigen_vals)
eigen_df$percent_variance = eigen_df$Eigenvalue/sum(eigen_df$Eigenvalue)

ggplot(eigen_df[1:20, ], aes(x = PC, y = percent_variance)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  ggtitle("Top 20 Eigenvalues from PCA of Gene Expression Data") +
  xlab("Principal Component") +
  ylab("Percentage of Variance Explained")

PC_df = data.frame(pc_scores)
dim(PC_df)
PC_df[1:2,1:3]
table(rownames(PC_df) == meta$`Counts file`)

PC_df = cbind(PC_df, meta)

pvs = round(eigen_df$percent_variance[1:2]*100,1)

ggplot(PC_df, aes(x = PC1, y = PC2, shape=model_organ, color=ko_gene)) + 
  labs(color="KO gene", shape = "model system") + 
  geom_point(size=2, alpha=0.8) + scale_color_brewer(palette = "Dark2") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2]))

round(eigen_df$percent_variance[1:5], 3)
```

### Only consider those extra-embryonic primitive syncytial cells

```{r fig.width = 7, fig.height=3.5}
dim(cts_dat_n)
cts_dat_n[1:2,1:3]
stopifnot(all(rownames(cts_dat_n) == meta$file_id))

cts_dat_pr = cts_dat_n[meta$model_organ == "extra-embryonic primitive syncytial cells",]
dim(cts_dat_pr)

sample_pca = prcomp(cts_dat_pr, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2

eigen_df = data.frame(PC = 1:length(eigen_vals), Eigenvalue = eigen_vals)
eigen_df$percent_variance = eigen_df$Eigenvalue/sum(eigen_df$Eigenvalue)

ggplot(eigen_df[1:20, ], aes(x = PC, y = percent_variance)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  ggtitle("Top 20 Eigenvalues from PCA of Gene Expression Data") +
  xlab("Principal Component") +
  ylab("Percentage of Variance Explained")

PC_df = data.frame(pc_scores)
dim(PC_df)
PC_df[1:2,1:3]
PC_df$file_id = rownames(PC_df)
PC_df = merge(PC_df, meta)
dim(PC_df)

pvs = round(eigen_df$percent_variance[1:2]*100,1)

ggplot(PC_df, aes(x = PC1, y = PC2, color=ko_gene)) + 
  geom_point(size = 2, alpha=0.8) + scale_color_brewer(palette = "Dark2") + 
  labs(color="KO gene") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2]))

round(eigen_df$percent_variance[1:5], 3)

```

```{r}
gc()
sessionInfo()
```



