---
title: "Check the results of differential expression analysis for 2024-05 release JAX_RNAseq2_ExtraEmbryonic"
author: "Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Checking the results of DE analysis. 

Save the sets of DE genes out, to prepare for next step of gene set enrichment analysis. 

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

#dat_path = "~/research/data/MorPhiC/May-2024/JAX_RNAseq2_ExtraEmbryonic/"
dat_path = "../../MorPhiC/data/May-2024/JAX_RNAseq2_ExtraEmbryonic/"
dat_path = paste0(dat_path, "processed/release110-gencode44/Tables")
```


## Read in meta data
```{r fig.width = 3, fig.height=3}
meta = fread("data/JAX_RNAseq2_ExtraEmbryonic_meta_data.tsv", 
             data.table = FALSE)
dim(meta)
meta[1:2,]
names(meta)

table(meta$model_organ, meta$ko_gene)
table(meta$run_id, meta$ko_gene)
table(meta$ko_strategy, meta$ko_gene)
```

## Read in gene count data and filter genes

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

stopifnot(setequal(names(cts)[-1], meta$file_id))

meta = meta[match(names(cts)[-1], meta$file_id),]
table(names(cts)[-1] == meta$file_id)

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}
table(meta$model_system, meta$model_organ)

model_s = meta$model_system
get_q75 <- function(x){quantile(x, probs = 0.75)}

gene_info = data.frame(Name = cts$Name, 
                       t(apply(cts_dat, 1, tapply, model_s, min)), 
                       t(apply(cts_dat, 1, tapply, model_s, get_q75)))

names(gene_info)[2:5] = c("ExM_min", "PrS_min", "ExM_q75", "PrS_q75")
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$ExM_q75 > 0, gene_info$PrS_q75 > 0)

w2kp = which(gene_info$ExM_q75 > 0 | gene_info$PrS_q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]
```

## Read in gene annoation
```{r fig.width = 4, fig.height=3}
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

table(gene_info$Name %in% gene_anno$ensembl_gene_id)
# all ensembl gene IDs are contained in the annotation
setdiff(gene_info$Name, gene_anno$ensembl_gene_id)

gene_info = merge(gene_anno, gene_info, by.x="ensembl_gene_id", by.y = "Name", 
                  all.x = FALSE, all.y = TRUE)
dim(gene_info)
gene_info[1:2,]

gene_info[gene_info$ExM_min > 2e4, c("ExM_min", "hgnc_symbol")]
gene_info[gene_info$PrS_min > 2e4, c("PrS_min", "hgnc_symbol")]

gene_info$gene_symbol = gene_info$hgnc_symbol

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

wEns = which(gene_info$gene_symbol == "" | is.na(gene_info$gene_symbol))
gene_info$gene_symbol[wEns] = gene_info$ensembl_gene_id[wEns]

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

```

## Prepare for reading in DE results
```{r fig.width = 4, fig.height=3}

result_dir = "results/2024_05_JAX_RNAseq2_ExtraEmbryonic"
processed_output_dir = file.path(result_dir, "processed")

all_result_files = list.files(path=processed_output_dir, pattern=".tsv", 
                              all.files=TRUE, full.names=FALSE)
all_result_files = sort(all_result_files)

extract_score <- function(x){
  x_split = str_split(x, "_")[[1]]
  x_start = 6
  x_end = length(x_split)-1
  x_d_group = paste(x_split[x_start:(x_end-1)], collapse="_")
  x_gene = x_split[x_end]
  return(c(x_d_group, x_gene))
}

df_file_info = NULL

for (x in all_result_files){
  df_file_info = rbind(df_file_info, extract_score(x))
}

df_file_info = as.data.frame(df_file_info)
colnames(df_file_info) = c("DE_group", "gene")
df_file_info

```

## Compare the number of DE genes

Make an exception for the case of DE group "ExM_20231004_23-robson-011" and gene "MEIS2", since this case only has one knockout sample, and the results are not reliable. 

```{r fig.width = 4, fig.height=3}

items = paste(df_file_info$DE_group, 
              df_file_info$gene, sep="_")
items

fc0 = log2(1.5)
p0  = 0.05

gene_symbols = df_file_info$gene
gene_symbols

table(gene_symbols %in% gene_info$gene_symbol)

df = data.frame(KO = items, gene_symbol = gene_symbols, 
                CE_nDE = NA, KO_nDE = NA, PTC_nDE = NA, 
                CE_padj = NA, CE_log2FoldChange = NA, 
                KO_padj = NA, KO_log2FoldChange = NA, 
                PTC_padj = NA, PTC_log2FoldChange = NA)

df$Model = str_extract(df$KO, "^[a-zA-Z0-9]+")
df$Condition = "nor"

gene_ids = gene_info$ensembl_gene_id[match(gene_symbols, gene_info$gene_symbol)]

genes_w_multi_ensembl = names(which(table(gene_info$gene_symbol)>1))
genes_w_multi_ensembl

for(k in 1:length(items)){

  it_k = items[k]
  
  res = fread(file.path(processed_output_dir, 
                        sprintf("2024_05_JAX_RNAseq2_ExtraEmbryonic_%s_DEseq2.tsv", it_k)), 
              data.table = FALSE)
  
  gene_id = gene_ids[k]
  gene_name = gene_symbols[k]
  w_gene = which(res$gene_ID == gene_id)
  stopifnot(length(w_gene) == 1)
  stopifnot(!(gene_name%in%genes_w_multi_ensembl))
  
  for(ko1 in c("CE", "KO", "PTC")){
    
    if ((it_k=="ExM_20231004_23-robson-011_MEIS2") & (ko1=="KO")){
      
      df[k, "KO_nDE"] = NA
      df[k, "KO_log2FoldChange"] = NA    
      df[k, "KO_padj"] = NA
      
    }else{
    
      fc = paste0(gene_name, "_", ko1, "_log2FoldChange")
      padj = paste0(gene_name, "_", ko1, "_padj")
      col_name = paste0(ko1, "_nDE")
      df[k, col_name] = sum(abs(res[[fc]]) > fc0 & res[[padj]] < p0, na.rm = TRUE)
      
      col_name = paste0(ko1, "_log2FoldChange")
      df[k,col_name] = res[[fc]][w_gene]
      
      col_name = paste0(ko1, "_padj")
      df[k,col_name] = res[[padj]][w_gene]
    }
  }
}

df

ggplot(df, aes(x=(PTC_nDE), y=(CE_nDE), col=Model,
               label=gene_symbol)) + 
    geom_point() +
    geom_text_repel(show.legend = FALSE) + ggtitle("# of DE genes") + 
    labs(x="PTC", y="CE") + 
    geom_abline(intercept = 0, slope = 1, color = "orange", linetype = "dashed")

ggplot(df, aes(x=(PTC_nDE), y=(KO_nDE), col=Model, 
               label=gene_symbol)) + 
    geom_point() +   
    geom_text_repel(show.legend = FALSE) + ggtitle("# of DE genes") + 
    labs(x="PTC", y="KO") + 
    geom_abline(intercept = 0, slope = 1, color = "orange", linetype = "dashed")

ggplot(df, aes(x=CE_log2FoldChange, y=-log10(CE_padj), col=Model, 
               label=gene_symbol)) + 
    geom_point(size=2) +   
    geom_text_repel(show.legend = FALSE) + ggtitle("DE results of the KO genes, CE") + 
    labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
    geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
    geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
               linetype = "dashed")

ggplot(df, aes(x=KO_log2FoldChange, y=-log10(KO_padj), col=Model, 
               label=gene_symbol)) + 
    geom_point(size=2) +  
    geom_text_repel(show.legend = FALSE) + ggtitle("DE results of the KO genes, KO") + 
    labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
    geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
    geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
               linetype = "dashed")

ggplot(df, aes(x=PTC_log2FoldChange, y=-log10(PTC_padj), col=Model, 
               label=gene_symbol)) + 
    geom_point(size=2) +
    geom_text_repel(show.legend = FALSE) + ggtitle("DE results of the KO genes, PTC") + 
    labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
    geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
    geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
               linetype = "dashed")

```

## Generate volcano plots

Also output the DE gene sets for the gene set enrichment analysis. 

When generating volcano plots, do not plot for the case of DE group "ExM_20231004_23-robson-011" and gene "MEIS2", since this case only has one knockout sample, and the results are not reliable. 

```{r fig.width = 4, fig.height=3}

df_file_info

for(k in 1:length(items)){
  
  it_k = items[k]
  d_group = df_file_info$DE_group[k]
  gene_name = df_file_info$gene[k]
  
  # create a list to store DE gene set for the gene set enrichment analysis
  de_gene_sets = list() 
  
  print(it_k)
  
  res = fread(file.path(processed_output_dir, 
                        sprintf("2024_05_JAX_RNAseq2_ExtraEmbryonic_%s_DEseq2.tsv", it_k)), 
              data.table = FALSE)
  
  for (ko1 in c("CE", "KO", "PTC")){

    res_k = res[,c(1, grep(ko1, names(res)))]
    names(res_k) = gsub(paste0(gene_name, "_", ko1, "_"), "", names(res_k))

    ww_up = which((res_k$log2FoldChange > log2(1.5)) & (res_k$padj < 0.05))
    ww_down = which((res_k$log2FoldChange < -log2(1.5)) & (res_k$padj < 0.05))

    de_gene_sets[[paste0(gene_name, "_", ko1, "_up")]] = res_k[ww_up,]$gene_ID
    de_gene_sets[[paste0(gene_name, "_", ko1, "_down")]] = res_k[ww_down,]$gene_ID
    
    res_k$DE = "No"
    res_k$DE[ww_up] <- "Up"
    res_k$DE[ww_down] <- "Down"
    
    res_k$log2FoldChange[which(res_k$log2FoldChange > 3)] = 3
    res_k$log2FoldChange[which(res_k$log2FoldChange < -3)] = -3
  
    col2use = c("blue", "grey", "red")
    names(col2use) = c("Down", "No", "Up")
    
    res_k$delabel = NA
    res_k$gene_symbol = gene_info$gene_symbol[match(res_k$gene_ID, 
                                                    gene_info$ensembl_gene_id)]
    w_de = which(res_k$DE != "No")
    res_k$delabel[w_de] = res_k$gene_symbol[w_de]
  
    print(table(res_k$DE))
    
    if ((it_k!="ExM_20231004_23-robson-011_MEIS2")|(ko1!="KO")){
      
      g2 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                      col=DE, label=delabel)) + 
           geom_point() + ggtitle(paste0(d_group, "\n", gene_name, "_", ko1)) + 
           geom_text_repel(point.padding = 0.5,
                           min.segment.length = 0, 
                           size = 3,  # Adjust text size
                           box.padding = 0.5,
                           max.overlaps = 20, 
                           show.legend = FALSE) + 
           scale_color_manual(values=col2use)
      
      print(g2)
    
    }

  }
  
  # save out the DE gene lists for gene set enrichment analysis
  DE_set_output_dir = sprintf("%s/DE_gene_lists", result_dir)
                       
    
  if (!file.exists(DE_set_output_dir)){
    dir.create(DE_set_output_dir)
  }
  
  saveRDS(de_gene_sets, 
          sprintf("%s/2024_05_JAX_RNAseq2_ExtraEmbryonic_%s_DE_gene_list.rds", 
                  DE_set_output_dir, it_k))
  
}

```


```{r}
gc()
sessionInfo()
```

