---
title: "Differential expression analysis for 2024-06 release JAX_RNAseq4_ExEm_brain"
author: "Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Differential expression using the JAX_RNAseq4_ExEm_brain bulk RNAseq data from the June release.

There are two model systems in this dataset. 

There are samples from different days. 

There is only one knockout strategy: PTC.

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

path = "../../MorPhiC/data/June-2024/JAX_RNAseq4_ExEm_brain/"
dat_path = paste0(path, "processed/release110-gencode44/Tables")
```


## Read in meta data
```{r fig.width = 3, fig.height=3}
meta = fread("data/June_2024/JAX_RNAseq4_ExEm_brain_meta_data.tsv", 
             data.table = FALSE)
dim(meta)
meta[1:2,]
names(meta)

table(meta$clonal.label)
table(table(meta$clonal.label))
table(table(meta$library_preparation.label))
table(table(meta$label))
```

## Read in gene count data and filter genes

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

stopifnot(setequal(names(cts)[-1], meta$file_id))

meta = meta[match(names(cts)[-1], meta$file_id),]
table(names(cts)[-1] == meta$file_id)

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name

table(meta$model_organ, meta$timepoint_value)

unique_timepoints = unique(meta$timepoint_value)
unique_timepoints = sort(unique_timepoints)

xx = as.character(meta$timepoint_value)
xx = factor(xx, levels = as.character(unique_timepoints))
meta$timepoint = xx
```

### Read in gene annoation
```{r}
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

table(rownames(cts_dat) %in% gene_anno$ensembl_gene_id)
# all ensembl gene IDs are contained in the annotation
setdiff(rownames(cts_dat), gene_anno$ensembl_gene_id)
```


### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}
model_s = meta$model_system
table(model_s, useNA="ifany")

get_q75 <- function(x){quantile(x, probs = 0.75)}

# there are two levels for model_system
# the result from apply(cts_dat, 1, tapply, model_s, min) is a matrix
# do transpose to the results from apply(cts_dat, 1, tapply, model_s, min)
gene_info = data.frame(Name = cts$Name, 
                       t(apply(cts_dat, 1, tapply, model_s, min)), 
                       t(apply(cts_dat, 1, tapply, model_s, median)), 
                       t(apply(cts_dat, 1, tapply, model_s, get_q75)), 
                       t(apply(cts_dat, 1, tapply, model_s, max)))

dim(gene_info)
gene_info[1:2, ]
table(row.names(gene_info)==gene_info$Name)

names(gene_info)[2:9] = paste0(rep(c("CBO_", "PrS_"), times=4), 
                               rep(c("min", "median", "q75", "max"), each=2))
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$CBO_q75 > 0, gene_info$PrS_q75 > 0)

w2kp = which(gene_info$CBO_q75 > 0 | gene_info$PrS_q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]

meta$read_depth = colSums(cts_dat)/1e6
meta$q75 = apply(cts_dat, 2, quantile, probs = 0.75)

# there are more than one model systems here
# need to set shape using model_system
ggplot(meta, aes(x=read_depth, y=q75, color=ko_strategy, shape = model_system)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  scale_shape_manual(values = c(7, 16)) + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "KO type", shape = "Model") + 
  scale_color_brewer(palette = "Set1")
```

```{r fig.width = 4.5, fig.height=3}
meta$run_id_short = str_replace(meta$run_id, "^[^-]*-", "")

table(meta$model_system, meta$run_id_short)

ggplot(meta, aes(x=read_depth, y=q75, color=run_id_short, shape = model_system)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  scale_shape_manual(values = c(7, 16)) + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "Run ID", shape = "Model") + 
  scale_color_brewer(palette = "Set1")

```

## Check the effect of run id among WT

```{r fig.width = 4.5, fig.height=3.2}
sample2kp = which(meta$ko_gene == "WT")
cts_dat_m = cts_dat[,sample2kp]
meta_m    = meta[sample2kp,]
summary(meta_m$q75)
dim(cts_dat_m)

stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
table(meta_m$file_id == colnames(cts_dat_m), useNA="ifany")

q75 = apply(cts_dat_m, 1, quantile, probs=0.75)
cts_dat_n = t(cts_dat_m[q75 > 0,])
cts_dat_n = log(median(meta_m$q75)*cts_dat_n/meta_m$q75 + 1)
dim(cts_dat_n)

sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2
eigen_vals[1:5]/sum(eigen_vals)
pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
pvs

PC_df = data.frame(pc_scores)
PC_df = cbind(PC_df, meta_m)

# there are two levels for model_system here
# need to use model_system for the shape
gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=model_system,
                        color=run_id_short)) +
  geom_point(size=2.5, alpha=0.7) + 
  #scale_color_brewer(palette = "Dark2") +
  scale_shape_manual(values = c(7, 16)) + 
  labs(color="Run id", shape ="Model") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("Wild type samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) + 
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)

table(PC_df$run_id, PC_df$timepoint)

gPC_time = ggplot(PC_df, aes(x = PC1, y = PC2, shape=model_system,
                            color=timepoint)) +
  geom_point(size=2.5, alpha=0.7) + 
  #scale_color_brewer(palette = "Dark2") +
  scale_shape_manual(values = c(7, 16)) + 
  labs(color="Day", shape ="Model") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("Wild type samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) + 
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC_time)
```

## Analyze samples of different model systems or batches

Explore the PC plots first, to decide running the models on what level of DE group. 

Then run the DE analysis. 

```{r fig.width = 4.5, fig.height=3.2, message=FALSE, warning=FALSE}
table(meta$run_id, meta$model_system)
table(meta$run_id, meta$ko_gene)
table(meta$ko_strategy, meta$ko_gene)

unique_model_ss = unique(meta$model_system)
unique_model_ss = sort(unique_model_ss)

for(model1 in unique_model_ss){
  print("")
  print("==========================================")
  print(model1)
  print("==========================================")
  print("")
  
  sample2kp = which(meta$model_system == model1)
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
  
  # verify the consistency between ko_strategy from metadata column and 
  # that from filenames
  extracted_ko_strings = str_extract_all(meta_m$file_id, "PTC|WT")
  print(table(sapply(extracted_ko_strings, length)))
  print(table(meta_m$ko_strategy, unlist(extracted_ko_strings)))
  
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  

  q75 = apply(cts_dat_m, 1, quantile, probs=0.75)

  cts_dat_n = t(cts_dat_m[q75 > 0,])
  cts_dat_n = log(median(meta_m$q75)*cts_dat_n/meta_m$q75 + 1)
  dim(cts_dat_n)
  
  sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
  names(sample_pca)
  pc_scores = sample_pca$x
  eigen_vals = sample_pca$sdev^2
  eigen_vals[1:5]/sum(eigen_vals)
  pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
  pvs
  
  PC_df = data.frame(pc_scores)
  PC_df = cbind(PC_df, meta_m)
  
  colnames(PC_df)[which(colnames(PC_df)=="timepoint")] = "day"
  
  only_ko_strategy = setdiff(unique(PC_df$ko_strategy), "WT")

  unique_ko_genes = sort(unique(PC_df$ko_gene))
  unique_ko_genes = c(unique_ko_genes[unique_ko_genes!="WT"], "WT")
  PC_df$ko_gene = factor(PC_df$ko_gene, unique_ko_genes)
  
  shapes = c(1, 7, 8, 9, 10, 11, 15, 16, 17)
  
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=ko_gene, color=day)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
    scale_shape_manual(values = shapes[1:length(unique(PC_df$ko_gene))]) + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle(paste0(model1, " ", only_ko_strategy, " for knockout genes")) + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) + 
  theme_classic() + 
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

  print(gPC)

  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=run_id_short, color=day)) +
  geom_point(size=2.5, alpha=0.7) + 
  scale_color_brewer(palette = "Dark2") +
  scale_shape_manual(values = shapes[1:length(unique(PC_df$run_id_short))]) + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle(paste0(model1, " ", only_ko_strategy, " for knockout genes")) + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme_classic() + 
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

  print(gPC)
  
  
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=run_id_short, color=ko_gene)) +
  geom_point(size=2.5, alpha=0.7) + 
  scale_color_brewer(palette = "Dark2") +
  scale_shape_manual(values = shapes[1:length(unique(PC_df$run_id_short))]) + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle(paste0(model1, " ", only_ko_strategy, " for knockout genes")) + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

  print(gPC)

}
```

Use model system + day or model system + day + run_id to construct DE group.

Exclude CBO day 32, because it only has one WT sample. 

Run DESeq2 for different DE groups separately
    
      CBO, day 9 only
      PrS, day 6, for two run_ids separately

For the output files, create two versions, one with direct information, one with padj set to NA for the genes when the number of 0 per test is >=4. 

In more details, for both versions of the outputs, make it one file per knockout gene per DE group (DE group can be model system, model system + condition, model system + run_id, model system + day, etc.). 

The first version of the output files contains 
    
    gene id, gene symbol

and for each knockout strategy:

    baseMean, log2FoldChange, lfcSE, stat, pvalue, padj
    
The second version of the output files contains  

    gene id, gene symbol, chr, strand, position

and for each knockout strategy:

    log2FoldChange, pvalue, padj (set to be NA if the number of 0 per test >= 4), 

In addition, save out a separate file of the sample size for each comparison. This needs to be by DE group. 

### Prepare output directories for DE results

```{r fig.width = 4.5, fig.height=3.2, message=FALSE, warning=FALSE}

df_size = NULL

release_name = "2024_06_JAX_RNAseq4_ExEm_brain"

output_dir = file.path("results", release_name)
raw_output_dir = file.path(output_dir, "raw")
processed_output_dir = file.path(output_dir, "processed")

if (!file.exists(raw_output_dir)){
  dir.create(raw_output_dir, recursive = TRUE)
}

if (!file.exists(processed_output_dir)){
  dir.create(processed_output_dir, recursive = TRUE)
}
```

### DE analysis for all genes

```{r fig.width = 4.5, fig.height=3.2, message=FALSE, warning=FALSE}
for(model1 in unique_model_ss){
  
  print("")
  print("==========================================")
  print(model1)
  print("==========================================")
  print("")
  
  sample2kp = which(meta$model_system == model1)
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
 
  table(meta_m$ko_strategy)
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  
  ## Generate sample information matrix
  cols2kp = c("model_system", "ko_strategy", "ko_gene", "run_id", "run_id_short", 
              "timepoint", "q75")
  sample_info = meta_m[,cols2kp]
  colnames(sample_info)[which(colnames(sample_info)=="timepoint")] = "day"
  
  dim(sample_info)
  sample_info[1:2,]
  
  sample_info$ko_gene_run_id = paste(sample_info$ko_gene, 
                                      sample_info$run_id_short, sep="_")
  
  print(table(sample_info$run_id, sample_info$ko_gene))
  print(table(sample_info$ko_gene, sample_info$day)) 
  print(table(sample_info$ko_gene_run_id, sample_info$day)) 
  
  # print the table of ko_gene and run_id by day
  unique_timepoints_model_s = unique(sample_info$day)
  unique_timepoints_model_s = as.character(sort(unique_timepoints_model_s))
  
  for (cur_day in as.character(unique_timepoints_model_s)){
    sample_info_by_day = sample_info[which(sample_info$day==as.character(cur_day)), ]
    print("---------------------------")
    print(paste0("Model system: ", model1))
    print(paste0("On day ", as.character(cur_day), ": "))
    print(table(sample_info_by_day$ko_gene, sample_info_by_day$run_id_short))     
  }
  
  # run DESeq2 for different day groups separately
  # For model system CBO:
  # day 9
  # day 32 (cannot do it, since there is only one WT sample)
  # 
  # For model system PrS:
  # day 6
  # but need to divide the two batches

  sample_info$model_day = paste0(sample_info$model_system, 
                              "_day_", 
                              as.character(sample_info$day))
  
  sample_info$de_grp = sample_info$model_day
  
  if(model1 == "PrS"){ 
    sample_info$de_grp = paste0(sample_info$de_grp, "_", 
                                sample_info$run_id)
  }
  
  unique_de_grps = unique(sample_info$de_grp)
  sorted_de_grps = sort(unique_de_grps)
  
  if(model1 == "CBO"){ 
      sorted_de_grps = rev(sorted_de_grps)
  }
  
  for (d_group in sorted_de_grps){

    dg_index = which(sample_info$de_grp==d_group)
    cts_dat_m_dg = cts_dat_m[, dg_index]
    sample_info_dg = sample_info[dg_index, ]
    
    stopifnot(all(sample_info_dg$de_grp==d_group))
    
    print("")
    print("-----------------------------------")
    print(paste0("Model system: ", model1))  
    if (sum(sample_info_dg$ko_gene=="WT") < 2){
      print(sprintf("DE group %s has only %d wild type samples", 
                    d_group, sum(sample_info_dg$ko_gene=="WT")))
      print(sprintf("do not run DESeq2 for it"))
      print("-----------------------------------")  
      next
    }else{
      print(sprintf("DE group %s DESeq2 results:", d_group))
      print("-----------------------------------")  
    }    
    
    
    # do two steps of filtering
    # first, filter based on q75 of each gene
    # second, filter based on whether each gene is expressed in at least 4 cells
    
    dg_q75 = apply(cts_dat_m_dg, 1, quantile, probs=0.75)
    cts_dat_m_dg = cts_dat_m_dg[dg_q75 > 0,]
    
    print(sprintf("After filtering by gene expression q75, the number of genes reduces from %d to %d", 
                  length(dg_q75), nrow(cts_dat_m_dg)))

    n_pos = rowSums(cts_dat_m_dg>0)
    cts_dat_m_dg = cts_dat_m_dg[n_pos >= 4,]

    if (length(n_pos)==nrow(cts_dat_m_dg)){
      print("After filtering by nonzero expression in at least 4 samples, the number of genes does not change.")
    }else{
      print(sprintf("After filtering by nonzero expression in at least 4 samples, the number of genes reduces from %d to %d", 
                    length(n_pos), nrow(cts_dat_m_dg)))    
    }
    
    # update the q75 based on only genes that will be used in the DE analysis
    sample_info_dg$q75 = apply(cts_dat_m_dg, 2, quantile, probs = 0.75)
    sample_info_dg$log_q75 = log(sample_info_dg$q75)
    
    if (length(unique(sample_info_dg$run_id))==1){
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene + log_q75)
    }else{
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene + run_id + log_q75)     
    }
  
    start.time <- Sys.time()
    dds = DESeq(dds)
    end.time <- Sys.time()

    time.taken <- end.time - start.time
    print(time.taken)
    
    ## association with read-depth
    res_rd = results(dds, name = "log_q75")
    res_rd = as.data.frame(res_rd)
    dim(res_rd)
    res_rd[1:2,]
    
    table(is.na(res_rd$padj))
  
    g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
      geom_histogram(color="darkblue", fill="lightblue", 
                     breaks = seq(0,1,by=0.01)) + 
      ggtitle(paste0(d_group, " read depth"))
    print(g0)
  
    ## Rerun the analysis without read-depth if it is not significant for 
    ## most genes, and the number of samples is small
    ## i.e., proportion of non-null in the distribution is smaller than 0.01
    ## (this needs to be restricted to the genes with not NA adjusted pvalue)
    ## or if smaller than 0.1 and the number of samples is not greater than 6

    pi_1_rd = max(0, mean(res_rd$pvalue[!is.na(res_rd$padj)] < 0.05) - 0.05)
    pi_1_rd = max(pi_1_rd, 0, 1 - 2*mean(res_rd$pvalue[!is.na(res_rd$padj)] > 0.5))
    
    print(sprintf("prop of non-null for rd: %.2e", pi_1_rd))
    
    if(pi_1_rd < 0.01 || (ncol(cts_dat_m_dg) <= 6 && pi_1_rd < 0.1 )){
      
      print("rerun DESeq2 without read depth")
      
      include_rd = 0
      
      if (length(unique(sample_info_dg$run_id))==1){
        dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                      ~ ko_gene)
      }else{
        dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                      ~ ko_gene + run_id)     
      }
      
      dds = DESeq(dds)
      
    }else{
      include_rd = 1
    }

    ## association with run_id
      
    if (length(unique(sample_info_dg$run_id))>1){
      if(include_rd)
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene + log_q75)
      else
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene)
      
      res_lrt = results(dds_lrt)
    
      res_run_id = as.data.frame(res_lrt)
      dim(res_run_id)
      res_run_id[1:2,]
    
      table(is.na(res_run_id$padj))
      
      g0 = ggplot(res_run_id %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, " Run ID"))
      print(g0)
    
    }
    
    ## DE analysis for each KO gene
    table(sample_info_dg$ko_gene)
    
    genos = setdiff(unique(sample_info_dg$ko_gene), "WT")
    genos
    
    for(geno in genos){
      
      res_g1 = results(dds, contrast = c("ko_gene", geno, "WT"))
      res_g1 = signif(data.frame(res_g1), 3)

      # add a column to record the number of zeros per test
      test_index = which(sample_info_dg$ko_gene%in%c(geno, "WT"))
      
      cts_dat_m_dg_test = cts_dat_m_dg[, test_index]
      n_zero = rowSums(cts_dat_m_dg_test==0)
      res_g1$n_zeros = n_zero    

      # record the number of samples involved in the comparison
      test_ko_gene_vec = sample_info_dg$ko_gene[test_index]
      
      df_size = rbind(df_size, 
                      c(d_group, 
                        paste0(geno, "_PTC"), 
                        sum(test_ko_gene_vec!="WT"), 
                        sum(test_ko_gene_vec=="WT")))   
      
      # prepare a processed version of output
      res_g1_reduced = res_g1[, c("log2FoldChange", "pvalue", "padj", "n_zeros")]
      res_g1_reduced$padj[which(res_g1_reduced$n_zeros>=4)] = NA
      
      if ((sum(test_ko_gene_vec!="WT")==1)|(sum(test_ko_gene_vec=="WT")==1)){
        res_g1_reduced$padj = NA
      }
      
      res_g1_reduced = res_g1_reduced[, c("log2FoldChange", "pvalue", "padj")]      
  
      n_DE = sum(res_g1_reduced$padj < 0.05 & abs(res_g1_reduced$log2FoldChange) > log2(1.5), 
                 na.rm = TRUE)
      print(paste0(d_group, " under KO gene ", geno, ":"))
      print(sprintf(" number of DE genes: %d", n_DE))

      g1 = ggplot(res_g1_reduced %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, " ", geno))
      print(g1)

      tag1 = sprintf("%s_PTC_", geno)
      colnames(res_g1) = paste0(tag1, colnames(res_g1))
      colnames(res_g1_reduced) = paste0(tag1, colnames(res_g1_reduced))
      
      
      res_df = data.frame(gene_ID=rownames(res_g1), 
                          res_g1)
      dim(res_df)
      res_df[1:2,]
      
      
      res_reduced_gene_anno = gene_anno[match(rownames(res_g1_reduced), 
                                              gene_anno$ensembl_gene_id), ]
      stopifnot(all(rownames(res_g1_reduced)==res_reduced_gene_anno$ensembl_gene_id))     
      
      # set gene hgnc_symbol that equal "" to NA
      hgnc_symbol_vec = res_reduced_gene_anno$hgnc_symbol
      hgnc_symbol_vec[which(hgnc_symbol_vec=="")] = NA
      
      res_reduced_df = data.frame(gene_ID=rownames(res_g1_reduced), 
                                  hgnc_symbol=hgnc_symbol_vec,
                                  chr=res_reduced_gene_anno$chr, 
                                  strand=res_reduced_gene_anno$strand, 
                                  start=res_reduced_gene_anno$start, 
                                  end=res_reduced_gene_anno$end, 
                                  res_g1_reduced)
      
      print("hgnc_symbol empty string and NA tables:")
      print(table(res_reduced_df$hgnc_symbol=="", useNA="ifany"))
      print(table(is.na(res_reduced_df$hgnc_symbol)))
        
      dim(res_reduced_df)
      res_reduced_df[1:2,]      
      
      setting_name = unique(sample_info_dg$model_day)
      
      fwrite(res_df, 
             sprintf("%s/%s_%s_%s_DEseq2_raw.tsv", 
                     raw_output_dir, release_name, setting_name, geno), 
             sep="\t")
      fwrite(res_reduced_df, 
             sprintf("%s/%s_%s_%s_DEseq2.tsv", 
                     processed_output_dir, release_name, setting_name, geno), 
             sep="\t")      

    }
    
  }
  
}
```

Save the size dataframe out
```{r fig.width = 4.5, fig.height=3.2, message=FALSE, warning=FALSE}

colnames(df_size) = c("DE_group", "knockout_gene_strategy", "n_ko", "n_WT")
df_size = as.data.frame(df_size)

multi_rows = which(table(df_size$knockout_gene_strategy)>1)
multi_rows

df_size[which(df_size$knockout_gene_strategy%in%names(multi_rows)),]

fwrite(df_size, 
       sprintf("%s/%s_DE_n_samples.tsv", output_dir, release_name), 
       sep="\t")


```

```{r}
gc()
sessionInfo()
```

