---
title: "Check the results of differential expression analysis for 2025-06 release JAX_RNAseq10"
author: "Bo Yu, Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

path = "../../MorPhiC/data/MorPhiC_internal_releases/MorPhiC_Release_June_2025/JAX_RNAseq10/"
dat_path = file.path(path, "processed/Tables")
```


## Read in meta data

```{r fig.width = 3, fig.height=3}
meta = fread("data/June_2025/JAX_RNAseq10_meta_data.tsv", data.table = FALSE)
dim(meta)
meta[1:2,]
names(meta)

table(meta$model_organ, meta$ko_gene)
table(meta$run_id, meta$ko_gene)
table(meta$ko_strategy, meta$ko_gene)

table(meta$ko_gene, meta$timepoint_value)
```

## Read in gene count data and filter genes

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

stopifnot(setequal(names(cts)[-1], meta$file_id))

meta = meta[match(names(cts)[-1], meta$file_id),]
table(names(cts)[-1] == meta$file_id, useNA="ifany")

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}

model_s = meta$model_system
table(model_s, useNA="ifany")

get_q75 <- function(x){quantile(x, probs = 0.75)}

# there are two levels for model_system
# the result from apply(cts_dat, 1, tapply, model_s, min) is a matrix
# do transpose to the results from apply(cts_dat, 1, tapply, model_s, min)
gene_info = data.frame(Name = cts$Name, 
                       t(apply(cts_dat, 1, tapply, model_s, min)), 
                       t(apply(cts_dat, 1, tapply, model_s, median)), 
                       t(apply(cts_dat, 1, tapply, model_s, get_q75)), 
                       t(apply(cts_dat, 1, tapply, model_s, max)))

dim(gene_info)
gene_info[1:2, ]
table(row.names(gene_info)==gene_info$Name, useNA="ifany")

names(gene_info)[2:9] = paste0(rep(c("ENDO_", "MESO_"), times=4), 
                               rep(c("min", "median", "q75", "max"), each=2))
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$ENDO_q75 > 0, gene_info$MESO_q75 > 0)

w2kp = which(gene_info$ENDO_q75 > 0 | gene_info$MESO_q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]
```

## Read in gene annotation
```{r fig.width = 4, fig.height=3}
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

# all genes are contained in the annotation file 
table(gene_info$Name %in% gene_anno$ensembl_gene_id)
setdiff(gene_info$Name, gene_anno$ensembl_gene_id)

gene_info = merge(gene_anno, gene_info, by.x="ensembl_gene_id", by.y = "Name", 
                  all.x = FALSE, all.y = TRUE)
dim(gene_info)
gene_info[1:2,]

gene_info[which(!gene_info$ensembl_gene_id%in%gene_anno$ensembl_gene_id)]

gene_info[gene_info$ENDO_min > 2e4, c("ENDO_min", "MESO_min", "hgnc_symbol")]
gene_info[gene_info$MESO_min > 2e4, c("ENDO_min", "MESO_min", "hgnc_symbol")]

gene_info$gene_symbol = gene_info$hgnc_symbol

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

wEns = which(gene_info$gene_symbol == "" | is.na(gene_info$gene_symbol))
gene_info$gene_symbol[wEns] = gene_info$ensembl_gene_id[wEns]

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

```

## Prepare for reading in DE results

This dataset has two model systems: ENDO and MESO, and one condition. 

DE analysis was run for different DE groups separately.
    
The specific DE groups (for example, run_id, knockout gene group, etc.) can be found in the file with naming in the format:

    ${dataset_name}_DE_n_samples.tsv
    

```{r fig.width = 4, fig.height=3}

release_name = "2025_06_JAX_RNAseq10"
result_dir = file.path("results", release_name)
processed_output_dir = file.path(result_dir, "processed")

all_result_files = list.files(path=processed_output_dir, pattern=".tsv", 
                              all.files=TRUE, full.names=FALSE)
all_result_files = sort(all_result_files)

regular_DE_files = all_result_files[!grepl("hyp_vs_nor", all_result_files)]
extract_cores <- function(x){
  x_split = str_split(x, "_")[[1]]
  x_start = 5
  x_end = length(x_split)-1
  x_d_group = paste(x_split[x_start:(x_end-1)], collapse="_")
  x_gene = x_split[x_end]
  return(c(x_d_group, x_gene))
}

df_file_info = NULL

for (x in regular_DE_files){
  df_file_info = rbind(df_file_info, extract_cores(x))
}

df_file_info = as.data.frame(df_file_info)
colnames(df_file_info) = c("DE_general_group", "gene")
df_file_info


```


## Prepare output directories for saving figure files out

```{r fig.width = 4.5, fig.height=3, message=FALSE, warning=FALSE}

df_size = NULL

figure_dir = file.path("figures", release_name)
volcano_figure_dir = file.path(figure_dir, "volcano_plots")

if (!file.exists(volcano_figure_dir)){
  dir.create(volcano_figure_dir, recursive = TRUE)
}

```



## Number of DE genes and properties of knockout gene

### For DE between knockout and WT samples

Plot the knockout genes from different DE groups all together in one figure. 

```{r fig.width =9, fig.height=3}
  
df_file_info$items = paste(df_file_info$DE_general_group, 
                           df_file_info$gene, sep="_")
df_file_info$items

fc0 = log2(1.5)
p0  = 0.05

gene_symbols = df_file_info$gene
gene_symbols

table(gene_symbols %in% gene_info$gene_symbol)

genes_w_multi_ensembl = names(which(table(gene_info$gene_symbol)>1))
genes_w_multi_ensembl


df = data.frame(KO = df_file_info$items, 
                group = df_file_info$DE_general_group,
                gene_symbol = df_file_info$gene, 
                strategy = ifelse(df_file_info$gene == 'ISL1', 'PTC', 'KO'),
                nDE = NA, 
                padj = NA, 
                log2FoldChange = NA)

df$Model = str_extract(df$KO, "^[a-zA-Z0-9]+")
  
gene_ids = gene_info$ensembl_gene_id[match(df_file_info$gene, gene_info$gene_symbol)]
gene_ids
    
for (k in 1:nrow(df_file_info)){
    
  res = fread(file.path(processed_output_dir,
                        paste0(release_name, "_", df_file_info$items[k], "_DEseq2.tsv")),
                        data.table = FALSE)
  names(res) = gsub("NKX2\\.1", "NKX2-1", names(res))
  print(res[1:2,])
  
  gene_id = gene_ids[k]
  w_gene = which(res$gene_ID == gene_id)
  
  stopifnot(length(w_gene) == 1)
  stopifnot(!(df$gene_symbol[k]%in%genes_w_multi_ensembl))
  
  fc = paste0(df$gene_symbol[k], '_', df$strategy[k], "_log2FoldChange")
  padj = paste0(df$gene_symbol[k], '_', df$strategy[k], "_padj")    

  df[k,"nDE"] = sum(abs(res[[fc]]) > fc0 & res[[padj]] < p0, na.rm = TRUE)
  df[k,"log2FoldChange"] = res[[fc]][w_gene]
  df[k,"padj"] = res[[padj]][w_gene]
  
}
  
df

p1 <- ggplot(df, aes(x=log2FoldChange, y=-log10(padj), 
               label=gene_symbol, color=group, shape = strategy)) + 
      geom_point(size=2) + scale_color_brewer(palette = "Set3")+ 
      geom_text_repel(point.padding = 0.5, min.segment.length = 0, 
                      size = 3,  # Adjust text size
                      box.padding = 0.5, 
                      max.overlaps = 20, show.legend = FALSE) + 
      #coord_cartesian(clip = "off") + 
      #ylim(0, max(-log10(df$PTC_padj)) + 0.5) + 
      ggtitle("DE results of the KO genes") + 
      labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
      geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
      geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
                 linetype = "dashed")

p2 <- ggplot(df, aes(x=log2FoldChange, y=nDE, 
               label=gene_symbol, color=group, shape = strategy)) + 
      geom_point(size=2) + scale_color_brewer(palette = "Set3")+ 
      geom_text_repel(point.padding = 0.5, min.segment.length = 0, 
                      size = 3,  # Adjust text size
                      box.padding = 0.5, 
                      max.overlaps = 20, show.legend = FALSE) + 
      #coord_cartesian(clip = "off") + 
      #ylim(0, max(-log10(df$PTC_padj)) + 0.5) + 
      ggtitle("DE results of the KO genes") + 
      labs(x="log2 fold change", y="number of DE genes")  + 
      geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
      geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
                 linetype = "dashed")

ggarrange(p1, p2, ncol = 2, nrow = 1)


```


## Generate volcano plots

In some settings, there are too many DE genes and they cannot all be labeled in the volcano plots.

### For DE between knockout and WT samples

```{r fig.width = 6.5, fig.height=5}

df_file_info

fc0
p0

for(k in 1:nrow(df_file_info)){
  
  it_k = df_file_info$items[k]
  d_group = df_file_info$DE_general_group[k]
  gene_name = df_file_info$gene[k]
  strategy = ifelse(gene_name == 'ISL1', 'PTC', 'KO')
  
  print(it_k)
  
  res = fread(file.path(processed_output_dir,
                        paste0(release_name, "_", it_k, "_DEseq2.tsv")),
              data.table = FALSE)
  names(res) = gsub("NKX2\\.1", "NKX2-1", names(res))

  res_k = res[,c(1, grep(strategy, names(res)))]
  
  names(res_k) = gsub(paste0(gene_name, "_", strategy, "_"), "", names(res_k))

  ww_up = which((res_k$log2FoldChange > fc0) & (res_k$padj < p0))
  ww_down = which((res_k$log2FoldChange < ((-1)*fc0)) & (res_k$padj < p0))
    
  res_k$DE = "No"
  res_k$DE[ww_up] <- "Up"
  res_k$DE[ww_down] <- "Down"
  
  if (any(res_k$log2FoldChange > 3)) res_k$log2FoldChange[which(res_k$log2FoldChange > 3)] = 3
  if (any(res_k$log2FoldChange < -3)) res_k$log2FoldChange[which(res_k$log2FoldChange < -3)] = -3

  col2use = c("blue", "grey", "red")
  names(col2use) = c("Down", "No", "Up")
    
  res_k$delabel = NA
  res_k$gene_symbol = gene_info$gene_symbol[match(res_k$gene_ID, 
                                                  gene_info$ensembl_gene_id)]
  w_de = which(res_k$DE != "No")
  res_k$delabel[w_de] = res_k$gene_symbol[w_de]

  print(table(res_k$DE))
  
  d_group_in_title = gsub("_nor", "", d_group)
  
  g2 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                    col=DE, label=delabel)) + 
         geom_point() + ggtitle(paste0(d_group_in_title, "\n", gene_name, "_", strategy)) + 
         geom_text_repel(point.padding = 0.5,
                         min.segment.length = 0, 
                         size = 3,  # Adjust text size
                         box.padding = 0.5,
                         max.overlaps = 20, 
                         show.legend = FALSE) + 
         scale_color_manual(values=col2use)
    
  print(g2)
  
  g3 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                    col=DE, label=delabel)) + 
         geom_point() + 
         geom_text_repel(point.padding = 0.5,
                         min.segment.length = 0, 
                         size = 3,  # Adjust text size
                         box.padding = 0.5,
                         max.overlaps = 20, 
                         show.legend = FALSE) + 
         scale_color_manual(values=col2use) + 
         theme(
          panel.background = element_rect(fill = "transparent", color = NA),
          plot.background = element_rect(fill = "transparent", color = NA),
          legend.background = element_rect(fill = "transparent", color = NA)
         )
  
  saved_figure_name = paste0(d_group_in_title, "_", gene_name, "_", strategy, ".svg")
  ggsave(file=file.path(volcano_figure_dir, saved_figure_name), 
         device = grDevices::svg,
         plot=g3, width=5, height=3.2, bg = "transparent", units="in")
}

```


```{r}
gc()
sessionInfo()
```

