---
title: "Check RNAseq data of 2024-09 release, JAX_RNAseq5_ExE_CBO"
author: "Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Check the bulk RNAseq JAX_RNAseq5_ExE_CBO data from September 2024 release.

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(readxl)

theme_set(theme_classic())

path = "../../MorPhiC/data/September-2024/JAX_RNAseq5_ExE_CBO"
dat_path = file.path(path, "processed/release110-gencode44/Tables")

read_header <- function(file_name, sheet){
  
  meta_header = read_excel(file_name, sheet = sheet, 
                         skip = 3, n_max = 1, col_names = FALSE)
  meta_header = as.character(meta_header)
  meta_header = gsub("cell_line.", "", meta_header, fixed = TRUE)
  meta_header = gsub("differentiated_cell_line.", "", meta_header, fixed = TRUE)
  meta_header = gsub("biomaterial_core.", "", meta_header, fixed = TRUE)
  meta_header = gsub("gene_expression_alteration_protocol.protocol_core.", 
                     "", meta_header, fixed = TRUE)
  meta_header = gsub(".text", "", meta_header, fixed = TRUE)
  
  meta_header = gsub("gene_expression_alteration_protocol.", "", 
                     meta_header, fixed = TRUE)

  meta_header
}

```

## Read in meta data

```{r fig.width = 3, fig.height=3}
meta_file = "../../MorPhiC/data/September-2024/JAX_RNAseq5_ExE_CBO/metadata/STUDY5_DPC_Metadata_template_v2.xlsx"

cl_header = read_header(meta_file, sheet = "Clonal cell line")
cl_header

cl = read_excel(meta_file, sheet = "Clonal cell line", skip = 5, col_names = FALSE)
names(cl) = cl_header
cl = as.data.frame(cl)
dim(cl)
cl[1:2,]
## drop the last three columns from the "Clonal cell line" tab since they don't have real content
cl = cl[, -c(8:10)]
dim(cl)
cl[1:2,]

dc_header = read_header(meta_file, sheet = "Differentiated product")
dc_header

dc = read_excel(meta_file, sheet = "Differentiated product", 
                skip = 5, col_names = FALSE)
names(dc) = dc_header
dc = as.data.frame(dc)
dim(dc)
dc[1:2,]
## drop the last two columns since they are columns with no content
dc = dc[, -c(9:10)]

table(dc$differentiated_timepoint_value, useNA="ifany")
## for this dataset, no need to manually exclude column 3 through column 8
## since comparing the metadata table of 2024 06 JAX_RNAseq5_ExE_CBO, 
## in the current dataset metadata file, 
## the empty columns from column 3 to column 8 are already excluded 
## dc = dc[,-(3:8)]


eas_header = read_header(meta_file, sheet = "Expression alteration strategy")
eas_header
eas_header = c("gene_expression_alteration_id", eas_header, 
               "altered_locus", "guide_sequence")
# need to manually change the column name for "protocol_id" in this case
# since the column "protocol_id" in tab "Clonal cell line" is no longer of the same format
# as in the content of "protocol_id" in "Expression alteration strategy" tab
eas_header[2] = "gene_expression_alteration_protocol.protocol_id"

eas = read_excel(meta_file, sheet = "Expression alteration strategy", 
                 skip = 5, col_names = FALSE)
names(eas) = eas_header
eas = as.data.frame(eas)
dim(eas)
eas[1:2,]
```

## Check cell lines

```{r}
any(duplicated(cl$biomaterial_id))
table(sapply(cl$protocol_id, function(x)str_split(x, "_")[[1]][1]))
table(cl$type)
table(cl$derived_cell_line_accession)
table(cl$zygosity)
sort(table(cl$clone_id), decreasing = TRUE)

table(dc$biomaterial_id %in% cl$biomaterial_id)
table(cl$biomaterial_id %in% dc$biomaterial_id)
setdiff(dc$biomaterial_id, cl$biomaterial_id)
setdiff(cl$biomaterial_id, dc$biomaterial_id)

table(table(dc$biomaterial_id))
table(dc$differentiated_timepoint_value)
table(dc$differentiated_terminally_differentiated)
table(dc$differentiation_protocol.protocol_core.protocol_id, 
      dc$model_organ)
table(dc$model_organ, 
      dc$differentiated_terminally_differentiated)
```

## Check Expression alteration strategy
```{r}
table(eas$gene_expression_alteration_protocol.protocol_id)
table(eas$allele_specific)
table(eas$gene_expression_alteration_protocol.protocol_id, 
      eas$altered_gene_symbols)
table(eas$gene_expression_alteration_protocol.protocol_id, 
      eas$crispr.sgrna_target)
```

## Extract KO strategy information

```{r}
dc[["ko_strategy"]] = 
  str_extract(dc$differentiated_biomaterial_description, 
              "\\(KO\\)|\\(CE\\)|\\(PTC\\)")
dc[["ko_strategy"]] = gsub("\\(|\\)", "", dc[["ko_strategy"]])

dc[["ko_gene"]] = str_extract(dc$differentiated_biomaterial_description, 
                                  "(?<=KOLF2.2 deleted ).\\S+")
table(dc$ko_strategy, dc$ko_gene, useNA = 'ifany')

table(dc$differentiated_biomaterial_description, 
      dc$ko_strategy, useNA = 'ifany')
```


## Read in library preparation information
```{r}
lib_header = read_header(meta_file, sheet = "Library preparation")
lib_header
lib_header = c(lib_header, 
               c("fragment_size", "input_amount", "input_unit", 
                 "final_yield", "final_yield_unit", 
                 "lib_concentration", "lib_concentration_unit",
                 "PCR_cycle", "PCR_cycle_sample_index"))

lib = read_excel(meta_file, sheet = "Library preparation", 
                skip = 5, col_names = FALSE)
dim(lib)
names(lib) = lib_header
lib = as.data.frame(lib)
dim(lib)
lib[1:2,]
```

## Read in sequence file information
```{r}
seq_header = read_header(meta_file, sheet = "Sequence file")
seq_header

seq = read_excel(meta_file, sheet = "Sequence file", 
                skip = 5, col_names = FALSE)
names(seq) = seq_header
seq = as.data.frame(seq)
seq$file_id = gsub("_R(1|2)_001.fastq.gz", "", 
                   seq$sequence_file.file_core.file_name)
dim(seq)
seq[1:2,]

names(seq)[which(names(seq)=="sequence_file.lane_index")] = "sequence_file.run_id"
names(seq)
  
seq = unique(seq[,c("file_id", "library_preparation.biomaterial_id", 
                    "sequencing_protocol.protocol_core.protocol_id", 
                    "sequence_file.run_id")])
dim(seq)
seq[1:2,]

table(seq$sequence_file.run_id)
```

## Merge tabs to make a flat metadata table
```{r}
intersect(names(dc), names(lib))
setequal(dc$differentiated_biomaterial_id, 
         lib$differentiated_biomaterial_id)
meta = merge(dc, lib)
dim(meta)
meta[1:2,]

intersect(names(meta), names(seq))
setequal(meta$library_preparation.biomaterial_id, 
         seq$library_preparation.biomaterial_id)
meta = merge(meta, seq)
dim(meta)
meta[1:2,]

update_name <- function(df1, old_name, new_name){
  ww1 = which(names(df1) == old_name)
  stopifnot(length(ww1) == 1)
  names(df1)[ww1] = new_name
  df1
}

meta = update_name(meta, "library_preparation.biomaterial_id", 
                   "lib_biomaterial_id")

meta = update_name(meta, "differentiation_protocol.protocol_core.protocol_id", 
                   "differentiation_protocol")

meta = update_name(meta, "library_preparation_protocol.protocol_core.protocol_id", 
                   "library_preparation_protocol")

meta = update_name(meta, "dissociation_protocol.protocol_core.protocol_id", 
                   "dissociation_protocol")

meta = update_name(meta, "sequencing_protocol.protocol_core.protocol_id", 
                   "sequencing_protocol")

meta = update_name(meta, "sequence_file.run_id", "run_id")

intersect(names(meta), names(cl))

meta = merge(meta, cl)
dim(meta)
meta[1:2,]

table(paste0("JAXPE003_", meta$biomaterial_id)==meta$protocol_id, 
      meta$protocol_id=="N.A")
      
table(meta$ko_strategy, useNA = "ifany")
meta$ko_strategy[which(is.na(meta$ko_strategy))] = "WT"

table(meta$ko_gene, useNA = "ifany")
meta$ko_gene[which(is.na(meta$ko_gene))] = "WT"

meta$model_system = NA
meta$model_system = str_extract(meta$differentiated_biomaterial_id, "^[^-]+")
table(meta$model_system, meta$model_organ)

table(meta$model_system, meta$ko_gene)

# verify consistency between ko_gene got here and those in the tab "Expression alteration strategy"
ko_gene_eas_matched = meta[match(eas$gene_expression_alteration_id, 
                                 meta$protocol_id), ]$ko_gene
table(ko_gene_eas_matched==eas$altered_gene_symbols, useNA="ifany")
  
fwrite(meta, file = "data/September_2024/JAX_RNAseq5_ExE_CBO_meta_data.tsv", sep="\t")
```

## Check count data

### genesCounts.csv

```{r fig.width = 5.5, fig.height=4}

cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

setdiff(names(cts)[-1], meta$file_id)
setdiff(meta$file_id, names(cts)[-1])
```


```{r}
gc()
sessionInfo()
```



