---
title: "Check the results of differential expression analysis for 2024_10 release NW_RNAseq_1"
author: "Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

path = "../../MorPhiC/data/October-2024/NW_RNAseq_1"
dat_path = file.path(path, "processed/release110-gencode44/Tables")
```


## Read in meta data

```{r fig.width = 3, fig.height=3}
meta = fread("data/October_2024/NW_RNAseq_1_meta_data.tsv", 
             data.table = FALSE)
dim(meta)
meta[1:2,]
names(meta)

meta$file_id
```

## Read in gene count data

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

stopifnot(setequal(names(cts)[-1], meta$file_id))

meta = meta[match(names(cts)[-1], meta$file_id),]
table(names(cts)[-1] == meta$file_id, useNA="ifany")

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}
get_q75 <- function(x){quantile(x, probs = 0.75)}

gene_info = data.frame(Name = cts$Name, 
                       apply(cts_dat, 1, min), 
                       apply(cts_dat, 1, median), 
                       apply(cts_dat, 1, get_q75), 
                       apply(cts_dat, 1, max))

dim(gene_info)
gene_info[1:2, ]
table(row.names(gene_info)==gene_info$Name, useNA="ifany")

names(gene_info)[2:5] = rep(c("min", "median", "q75", "max"), each=1)
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$q75 > 0)

w2kp = which(gene_info$q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]
```

## Read in gene annotation
```{r fig.width = 4, fig.height=3}
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

# all genes are contained in the annotation file 
table(gene_info$Name %in% gene_anno$ensembl_gene_id)
setdiff(gene_info$Name, gene_anno$ensembl_gene_id)

gene_info = merge(gene_anno, gene_info, by.x="ensembl_gene_id", by.y = "Name", 
                  all.x = FALSE, all.y = TRUE)
dim(gene_info)
gene_info[1:2,]

gene_info[which(!gene_info$ensembl_gene_id%in%gene_anno$ensembl_gene_id)]

gene_info[gene_info$min > 2e4, c("min", "hgnc_symbol")]


gene_info$gene_symbol = gene_info$hgnc_symbol

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

wEns = which(gene_info$gene_symbol == "" | is.na(gene_info$gene_symbol))
gene_info$gene_symbol[wEns] = gene_info$ensembl_gene_id[wEns]

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

```

## Prepare for reading in DE results

For the October 2024 release of NW_RNAseq_1, the knockout strategy is degron.

This dataset involves two genes and three treatment settings:

    DMSO
    Auxin_6h
    Auxin_24h

Each knockout gene is used as a DE group. 

The comparison is between each pair of two treatments. 

The specific DE groups can be found in the file with naming in the format:

    ${dataset_name}_DE_n_samples.tsv

```{r fig.width = 4, fig.height=3}

release_name = "2024_10_NW_RNAseq_1"
result_dir = file.path("results", release_name)
processed_output_dir = file.path(result_dir, "processed")

all_result_files = list.files(path=processed_output_dir, pattern=".tsv", 
                              all.files=TRUE, full.names=FALSE)
all_result_files = sort(all_result_files)

extract_cores <- function(x){
  x_split = str_split(x, "_")[[1]]
  x_start = 6
  x_end = length(x_split)-1
  x_d_group = x_split[x_start]
  x_pairs = paste(x_split[(x_start+1):(x_end)], collapse="_")
  return(c(x_d_group, x_pairs))
}

df_file_info = NULL

for (x in all_result_files){
  df_file_info = rbind(df_file_info, extract_cores(x))
}

df_file_info = as.data.frame(df_file_info)
colnames(df_file_info) = c("knockout_gene", "treatment_pair")
df_file_info

```

## Prepare output directories for saving figure files out

```{r fig.width = 4.5, fig.height=3, message=FALSE, warning=FALSE}

df_size = NULL

figure_dir = file.path("figures", release_name)
volcano_figure_dir = file.path(figure_dir, "volcano_plots")

if (!file.exists(volcano_figure_dir)){
  dir.create(volcano_figure_dir, recursive = TRUE)
}

```


## Number of DE genes and properties of knockout gene

Plot the treatment pairs under different knockout genes all together in one figure.

```{r fig.width =5, fig.height=3}

df_file_info$items = paste(df_file_info$knockout_gene, 
                           df_file_info$treatment_pair, sep="_")
df_file_info$items

table(table(df_file_info$items))

fc0 = log2(1.5)
p0  = 0.05

gene_symbols = df_file_info$knockout_gene
gene_symbols

table(gene_symbols %in% gene_info$gene_symbol)

genes_w_multi_ensembl = names(which(table(gene_info$gene_symbol)>1))
genes_w_multi_ensembl

  
df = data.frame(KO = df_file_info$items, 
                ko_gene = df_file_info$knockout_gene,
                treat_pair = df_file_info$treatment_pair, 
                nDE = NA, 
                padj = NA, 
                log2FoldChange = NA)

gene_ids = gene_info$ensembl_gene_id[match(df_file_info$knockout_gene, 
                                           gene_info$gene_symbol)]
gene_ids


for (k in 1:nrow(df_file_info)){
    
  res = fread(file.path(processed_output_dir,
                        paste0(release_name, "_", df_file_info$items[k], "_DEseq2.tsv")),
                        data.table = FALSE)
  print(res[1:2,])
  
  gene_id = gene_ids[k]
  w_gene = which(res$gene_ID == gene_id)

  stopifnot(length(w_gene) == 1)
  stopifnot(!(df$ko_gene[k]%in%genes_w_multi_ensembl))
  
  fc = paste0(df$treat_pair[k], "_log2FoldChange")
  padj = paste0(df$treat_pair[k], "_padj")

  df[k,"nDE"] = sum(abs(res[[fc]]) > fc0 & res[[padj]] < p0, na.rm = TRUE)
  df[k,"log2FoldChange"] = res[[fc]][w_gene]
  df[k,"padj"] = res[[padj]][w_gene]
  
}
  
df

p1 <- ggplot(df, aes(x=log2FoldChange, y=-log10(padj), 
               label=ko_gene, color=treat_pair)) + 
      geom_point(size=2) + scale_color_brewer(palette = "Dark2")+ 
      geom_text_repel(point.padding = 0.5, min.segment.length = 0, 
                      size = 3,  # Adjust text size
                      box.padding = 0.5, 
                      max.overlaps = 20, show.legend = FALSE) + 
      ggtitle("DE results of treatment pairs under each KO gene\nplotted for each KO gene itself") + 
      labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
      geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
      geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
                 linetype = "dashed")

print(p1)
```

```{r fig.width =4, fig.height=3.2}
p2 <- ggplot(df, aes(x=ko_gene, y=nDE, fill=treat_pair)) +
      geom_bar(position="dodge", stat="identity")+theme_classic() + 
      ylab("Number of DE genes")+
      ggtitle("Number of DE genes\nfrom each treatment pair\nunder each KO gene")

print(p2)

```


## Generate volcano plots

In some settings, there are too many DE genes and they cannot all be labeled in the volcano plots.

```{r fig.width = 6.5, fig.height=5}

df_file_info

fc0
p0

for(k in 1:nrow(df_file_info)){
  
  it_k = df_file_info$items[k]
  ko_gene = df_file_info$knockout_gene[k]
  trt_pair = df_file_info$treatment_pair[k]
  
  print(it_k)
  
  res = fread(file.path(processed_output_dir,
                        paste0(release_name, "_", it_k, "_DEseq2.tsv")),
              data.table = FALSE)


  res_k = res[,c(1, grep(trt_pair, names(res)))]
  names(res_k) = gsub(paste0(trt_pair, "_"), "", names(res_k))

  ww_up = which((res_k$log2FoldChange > fc0) & (res_k$padj < p0))
  ww_down = which((res_k$log2FoldChange < ((-1)*fc0)) & (res_k$padj < p0))
    
  res_k$DE = "No"
  res_k$DE[ww_up] <- "Up"
  res_k$DE[ww_down] <- "Down"
    
  res_k$log2FoldChange[which(res_k$log2FoldChange > 3)] = 3
  res_k$log2FoldChange[which(res_k$log2FoldChange < -3)] = -3

  col2use = c("blue", "grey", "red")
  names(col2use) = c("Down", "No", "Up")
    
  res_k$delabel = NA
  res_k$gene_symbol = gene_info$gene_symbol[match(res_k$gene_ID, 
                                                  gene_info$ensembl_gene_id)]
  w_de = which(res_k$DE != "No")
  res_k$delabel[w_de] = res_k$gene_symbol[w_de]

  print(table(res_k$DE))

  g2 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                    col=DE, label=delabel)) + 
         geom_point() + ggtitle(paste0("Under knockout gene ", ko_gene, "\n", trt_pair)) + 
         geom_text_repel(point.padding = 0.5,
                         min.segment.length = 0, 
                         size = 3,  # Adjust text size
                         box.padding = 0.5,
                         max.overlaps = 20, 
                         show.legend = FALSE) + 
         scale_color_manual(values=col2use)
    
  print(g2)
  
  g3 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                    col=DE, label=delabel)) + 
         geom_point() +
         geom_text_repel(point.padding = 0.5,
                         min.segment.length = 0, 
                         size = 3,  # Adjust text size
                         box.padding = 0.5,
                         max.overlaps = 20, 
                         show.legend = FALSE) + 
         scale_color_manual(values=col2use) + 
         theme(
          panel.background = element_rect(fill = "transparent", color = NA),
          plot.background = element_rect(fill = "transparent", color = NA),
          legend.background = element_rect(fill = "transparent", color = NA)
         )
  
  saved_figure_name = paste0(it_k, ".svg")
  ggsave(file=file.path(volcano_figure_dir, saved_figure_name), 
         plot=g3, width=5, height=3.2, bg = "transparent", units="in")
  
}


```



```{r}
gc()
sessionInfo()
```

