---
title: "Differential expression analysis for 2024-09 release UCSF_Bulk_EB_Timecourse"
author: "Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# explicitly set warning level in R, to prevent a warning from being converted to an error
options(warn=1)
```

PC plot using the bulk RNAseq UCSF_Bulk_EB_Timecourse data from the September 2024 release.

This dataset does not involve any knockout. 

There are different time points. 

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

path = "../../MorPhiC/data/September-2024/UCSF_Bulk_EB_Timecourse/"
dat_path = paste0(path, "processed/release110-gencode44/Tables")
```


## Read in meta data
```{r fig.width = 3, fig.height=3}
meta = fread("data/September_2024/UCSF_Bulk_EB_Timecourse_meta_data.tsv", 
             data.table = FALSE)
dim(meta)
meta[1:2,]
names(meta)

meta$file_id
```

### Extract biomaterial name, day, and biomaterial type
```{r}

biomaterial_name_column = rep(NA, nrow(meta))
biomaterial_type_column = rep(NA, nrow(meta))
day_column = rep(NA, nrow(meta))

for (i in 1:nrow(meta)){
  x = meta$file_id[i]
  x_substrings = str_split(x, "_")[[1]]
  biomaterial_type_column[i] = x_substrings[1]
  if (x_substrings[1]=="iPSC"){
    biomaterial_name_column[i] = paste(x_substrings[1:2], collapse="_")
    day_column[i] = "non"
  }else{
    biomaterial_name_column[i] = paste(x_substrings[1:4], collapse="_")
    day_column[i] = x_substrings[3]  
  }
}

meta$biomaterial_name = biomaterial_name_column
meta$biomaterial_type = biomaterial_type_column
meta$day = day_column

table(meta$day)
```
## Read in gene count data

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

stopifnot(setequal(names(cts)[-1], meta$file_id))

meta = meta[match(names(cts)[-1], meta$file_id),]
table(names(cts)[-1] == meta$file_id, useNA="ifany")

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name

table(meta$run_id, meta$day)

unique_timepoints_nonna = as.integer(setdiff(unique(meta$day), "non"))
unique_timepoints = c(as.character(sort(unique_timepoints_nonna)), "non")

meta$day = factor(meta$day, levels = as.character(unique_timepoints))

```

### Read in gene annoation
```{r}
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

table(rownames(cts_dat) %in% gene_anno$ensembl_gene_id)
# all ensembl gene IDs are contained in the annotation
setdiff(rownames(cts_dat), gene_anno$ensembl_gene_id)
```


### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4.5, fig.height=3}

get_q75 <- function(x){quantile(x, probs = 0.75)}

gene_info = data.frame(Name = cts$Name, 
                       apply(cts_dat, 1, min), 
                       apply(cts_dat, 1, median), 
                       apply(cts_dat, 1, get_q75), 
                       apply(cts_dat, 1, max))

dim(gene_info)
gene_info[1:2, ]
table(row.names(gene_info)==gene_info$Name, useNA="ifany")

names(gene_info)[2:5] = rep(c("min", "median", "q75", "max"), each=1)
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$q75 > 0)

w2kp = which(gene_info$q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]

meta$read_depth = colSums(cts_dat)/1e6
meta$q75 = apply(cts_dat, 2, quantile, probs = 0.75)

ggplot(meta, aes(x=read_depth, y=q75, color=day, shape = biomaterial_type)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  scale_shape_manual(values = c(7, 16)) + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "Day", shape = "biomaterial_type") + 
  scale_color_brewer(palette = "Dark2")
```

## Generate PC plot

```{r fig.width = 6, fig.height=4.5}
summary(meta$q75)
dim(cts_dat)

stopifnot(all(meta$file_id == colnames(cts_dat)))
table(meta$file_id == colnames(cts_dat), useNA="ifany")

cts_dat_n = t(cts_dat)
cts_dat_n = log(median(meta$q75)*cts_dat_n/meta$q75 + 1)
dim(cts_dat_n)

sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2
eigen_vals[1:5]/sum(eigen_vals)
pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
pvs

PC_df = data.frame(pc_scores)
PC_df = cbind(PC_df, meta)

gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=biomaterial_type, color=day)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
  labs(color="Day", shape ="Biomaterial_type") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("All samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)

```

## Analysis samples of different days

For the September 2024 release of UCSF_Bulk_EB_Timecourse, all samples are WT samples. 

Run DE analysis between samples of consecutive days, excluding iPSC samples.

### Prepare output directories for DE results

```{r message=FALSE, warning=FALSE}

release_name = "2024_09_UCSF_Bulk_EB_Timecourse"

output_dir = file.path("results", release_name)
raw_output_dir = file.path(output_dir, "raw")
processed_output_dir = file.path(output_dir, "processed")

if (!file.exists(raw_output_dir)){
  dir.create(raw_output_dir, recursive = TRUE)
}

if (!file.exists(processed_output_dir)){
  dir.create(processed_output_dir, recursive = TRUE)
}

```

### Run DE analysis for all pairs of consecutive days

```{r fig.width = 4.8, fig.height=3.2, message=FALSE, warning=FALSE}

df_size = NULL

unique_timepoints_nonna_chars = setdiff(unique_timepoints, "non")

for(i in 1:(length(unique_timepoints_nonna_chars)-1)){
  
  day_first = unique_timepoints_nonna_chars[i]
  day_second = unique_timepoints_nonna_chars[i+1]
  
  compare_string = paste0("day ", day_second, "vs day ", day_first)
  
  print("")
  print("==========================================")
  print(paste0("EB WT ", compare_string))
  print("==========================================")
  print("")
  
  sample2kp = which(meta$day%in%c(day_first, day_second))
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
 
  table(meta_m$day)
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  
  ## Generate sample information matrix
  cols2kp = c("run_id", "day")
  sample_info = meta_m[,cols2kp]
  
  dim(sample_info)
  sample_info[1:2,]
  
  print(table(sample_info$run_id, sample_info$day))

  d_group = paste0("day_", day_second, "_vs_day_", day_first)

  stopifnot(all(sample_info$day%in%c(day_first, day_second)))
    
  print("")
  print("-----------------------------------")
  print(paste0("Model system: ", "EB"))  
  print(sprintf("DE group %s DESeq2 results:", d_group))
  print("-----------------------------------")  
     
  # do two steps of filtering
  # first, filter based on q75 of each gene
  # second, filter based on whether each gene is expressed in at least 4 cells
    
  m_q75 = apply(cts_dat_m, 1, quantile, probs=0.75)
  cts_dat_m = cts_dat_m[m_q75 > 0,]
    
  print(sprintf("After filtering by gene expression q75, the number of genes reduces from %d to %d", 
                length(m_q75), nrow(cts_dat_m)))

  n_pos = rowSums(cts_dat_m>0)
  cts_dat_m = cts_dat_m[n_pos >= 4,]

  if (length(n_pos)==nrow(cts_dat_m)){
    print("After filtering by nonzero expression in at least 4 samples, the number of genes does not change.")
  }else{
    print(sprintf("After filtering by nonzero expression in at least 4 samples, the number of genes reduces from %d to %d", 
                  length(n_pos), nrow(cts_dat_m)))    
  }
    
  # update the q75 based on only genes that will be used in the DE analysis
  sample_info$q75 = apply(cts_dat_m, 2, quantile, probs = 0.75)
  sample_info$log_q75 = log(sample_info$q75)
    
  sample_info$day = as.factor(sample_info$day)
  
  dds = DESeqDataSetFromMatrix(cts_dat_m, colData=sample_info, 
                                ~ day + log_q75)

  
  start.time <- Sys.time()
  dds = DESeq(dds)
  end.time <- Sys.time()

  time.taken <- end.time - start.time
  print(time.taken)
  
  ## association with read-depth
  res_rd = results(dds, name = "log_q75")
  res_rd = as.data.frame(res_rd)
  dim(res_rd)
  res_rd[1:2,]
    
  table(is.na(res_rd$padj))

  g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
    geom_histogram(color="darkblue", fill="lightblue", 
                   breaks = seq(0,1,by=0.01)) + 
    ggtitle(paste0("EB WT ", d_group, " read depth"))
  print(g0)
  
  ## Rerun the analysis without read-depth if it is not significant for 
  ## most genes, and the number of samples is small
  ## i.e., proportion of non-null in the distribution is smaller than 0.01
  ## (this needs to be restricted to the genes with not NA adjusted pvalue)
  ## or if smaller than 0.1 and the number of samples is not greater than 6

  pi_1_rd = max(0, mean(res_rd$pvalue[!is.na(res_rd$padj)] < 0.05) - 0.05)
  pi_1_rd = max(pi_1_rd, 0, 1 - 2*mean(res_rd$pvalue[!is.na(res_rd$padj)] > 0.5))
  
  print(sprintf("prop of non-null for rd: %.2e", pi_1_rd))
  
  if(pi_1_rd < 0.01 || (ncol(cts_dat_m) <= 6 && pi_1_rd < 0.1 )){
    print("rerun DESeq2 without read depth")
    dds = DESeqDataSetFromMatrix(cts_dat_m, colData=sample_info, 
                                  ~ day)
    dds = DESeq(dds)
  }
    
  ## DE analysis between two days
  table(sample_info$day)
    
  res_g1 = results(dds, contrast = c("day", day_second, day_first))
  res_g1 = signif(data.frame(res_g1), 3)

  # add a column to record the number of zeros per test
  n_zero = rowSums(cts_dat_m==0)
  res_g1$n_zeros = n_zero    

  # record the number of samples involved in the comparison
  test_day_vec = sample_info$day
      
  df_size = rbind(df_size, 
                  c(paste0("EB_", d_group), 
                    "WT", 
                    sum(test_day_vec==day_second), 
                    sum(test_day_vec==day_first)))   
      
  # prepare a processed version of output
  res_g1_reduced = res_g1[, c("log2FoldChange", "pvalue", "padj", "n_zeros")]
  res_g1_reduced$padj[which(res_g1_reduced$n_zeros>=4)] = NA
  
  if ((sum(test_day_vec==day_second)==1)|(sum(test_day_vec==day_first)==1)){
    res_g1_reduced$padj = NA
  }
  
  res_g1_reduced = res_g1_reduced[, c("log2FoldChange", "pvalue", "padj")]      

  n_DE = sum(res_g1_reduced$padj < 0.05 & abs(res_g1_reduced$log2FoldChange) > log2(1.5), 
             na.rm = TRUE)
  print(paste0("EB WT ", d_group, ":"))
  print(sprintf("number of DE genes: %d", n_DE))

  g1 = ggplot(res_g1_reduced %>% subset(!is.na(padj)), aes(x=pvalue)) + 
    geom_histogram(color="darkblue", fill="lightblue", 
                   breaks = seq(0,1,by=0.01)) + 
    ggtitle(paste0("EB WT ", d_group))
  print(g1)

  tag1 = sprintf("%s_", d_group)
  colnames(res_g1) = paste0(tag1, colnames(res_g1))
  colnames(res_g1_reduced) = paste0(tag1, colnames(res_g1_reduced))
      
  res_df = data.frame(gene_ID=rownames(res_g1), 
                      res_g1)
  dim(res_df)
  res_df[1:2,]
      
  res_reduced_gene_anno = gene_anno[match(rownames(res_g1_reduced), 
                                          gene_anno$ensembl_gene_id), ]
  stopifnot(all(rownames(res_g1_reduced)==res_reduced_gene_anno$ensembl_gene_id))     
  
  # set gene hgnc_symbol that equal "" to NA
  hgnc_symbol_vec = res_reduced_gene_anno$hgnc_symbol
  hgnc_symbol_vec[which(hgnc_symbol_vec=="")] = NA
      
  res_reduced_df = data.frame(gene_ID=rownames(res_g1_reduced), 
                              hgnc_symbol=hgnc_symbol_vec,
                              chr=res_reduced_gene_anno$chr, 
                              strand=res_reduced_gene_anno$strand, 
                              start=res_reduced_gene_anno$start, 
                              end=res_reduced_gene_anno$end, 
                              res_g1_reduced)
  
  print("hgnc_symbol empty string and NA tables:")
  print(table(res_reduced_df$hgnc_symbol=="", useNA="ifany"))
  print(table(is.na(res_reduced_df$hgnc_symbol)))
        
  dim(res_reduced_df)
  res_reduced_df[1:2,]      
  
  setting_name = paste0("EB_WT_", d_group)

  fwrite(res_df, 
         sprintf("%s/%s_%s_DEseq2_raw.tsv", 
                 raw_output_dir, release_name, setting_name), 
         sep="\t")
  fwrite(res_reduced_df, 
         sprintf("%s/%s_%s_DEseq2.tsv", 
                 processed_output_dir, release_name, setting_name), 
         sep="\t")      
  
}

```

Save the size dataframe out
```{r fig.width = 4.5, fig.height=3.2, message=FALSE, warning=FALSE}

colnames(df_size) = c("DE_group", "knockout_gene_strategy", "n_day_second", "n_day_first")
df_size = as.data.frame(df_size)

fwrite(df_size, 
       sprintf("%s/%s_DE_n_samples.tsv", output_dir, release_name), 
       sep="\t")

```

```{r}
gc()
sessionInfo()
```

