---
title: "Check the results of differential expression analysis"
author: "Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

#dat_path = "~/research/data/MorPhiC/December-2023/JAX_RNAseq_ExtraEmbryonic/"
dat_path = "../../MorPhiC/data/December-2023/JAX_RNAseq_ExtraEmbryonic/"
dat_path = paste0(dat_path, "processed/Tables")
```


## Read in meta data
```{r fig.width = 3, fig.height=3}
meta_flat = read_excel("data/JAX_RNAseq_ExtraEmbryonic_meta_falttened.xlsx")
meta_flat = as.data.frame(meta_flat)
dim(meta_flat)
meta_flat[1:2,]
names(meta_flat)
```

### Gather cell line id information
```{r fig.width = 3, fig.height=3}
meta_flat$`DIFFERENTIATED CELL LINE ID`[1:5]
cell_line_id = str_split(meta_flat$`DIFFERENTIATED CELL LINE ID`, pattern = '-')
table(sapply(cell_line_id, length))

cell_line_id = lapply(cell_line_id, function(x) { length(x) <- 4; x})
cell_line_id = as.data.frame(do.call(rbind, cell_line_id))
dim(cell_line_id)
names(cell_line_id) = c("model_system", "cell_line", "clone_id", "condition")
cell_line_id[1:2,]
cell_line_id$condition[which(is.na(cell_line_id$condition))] = "nor"

table(meta_flat$`Model System`, cell_line_id$model_system)
table(cell_line_id$model_system, cell_line_id$condition, useNA="ifany")
```

### Add ```run_id``` and ```ko_strategy```. 

```{r fig.width = 3, fig.height=3}
w2update = which(meta_flat$`RUN ID` == "20231004_23-robson-008-run2")
meta_flat$run_id = meta_flat$`RUN ID`
meta_flat$run_id[w2update] = "20230918_23-robson-008"

table(meta_flat$run_id, cell_line_id$model_system, useNA="ifany")
table(meta_flat$run_id, cell_line_id$condition, useNA="ifany")

table(meta_flat$`TARGETED GENOMIC REGION`, useNA="ifany")
meta_flat$ko_strategy = str_extract(meta_flat$`TARGETED GENOMIC REGION`, 
                                    "\\(\\S+\\)$")
meta_flat$ko_strategy = str_remove_all(meta_flat$ko_strategy, "\\(|\\)")
meta_flat$ko_strategy[which(is.na(meta_flat$ko_strategy))] = "WT"

table(meta_flat$`TARGETED GENOMIC REGION`, meta_flat$ko_strategy, 
      useNA="ifany")

meta_flat$ko_gene = meta_flat$`ALTERED GENE SYMBOLS`
meta_flat$ko_gene[which(is.na(meta_flat$ko_gene))] = "WT"

table(cell_line_id$model_system, meta_flat$ko_gene)
table(meta_flat$run_id, meta_flat$ko_gene)
table(meta_flat$ko_strategy, meta_flat$ko_gene)
```

## Read in gene count data and filter genes

```{r fig.width = 4, fig.height=3}
meta = cbind(meta_flat, cell_line_id)
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

setequal(names(cts)[-1], meta$`Counts file`)

meta = meta[match(names(cts)[-1], meta$`Counts file`),]
table(names(cts)[-1] == meta$`Counts file`)

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}
model_s = meta$model_system
get_q75 <- function(x){quantile(x, probs = 0.75)}

gene_info = data.frame(Name = cts$Name, 
                       t(apply(cts_dat, 1, tapply, model_s, min)), 
                       t(apply(cts_dat, 1, tapply, model_s, get_q75)))
dim(gene_info)
gene_info[1:2,]

names(gene_info)[2:5] = c("ExM_min", "PrS_min", "ExM_q75", "PrS_q75")
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$ExM_q75 > 0, gene_info$PrS_q75 > 0)

w2kp = which(gene_info$ExM_q75 > 0 | gene_info$PrS_q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]
```

## Read in gene annoation
```{r fig.width = 4, fig.height=3}
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

table(gene_info$Name %in% gene_anno$ensembl_gene_id)
# all ensembl gene IDs are contained in the annotation
setdiff(gene_info$Name, gene_anno$ensembl_gene_id)

gene_info = merge(gene_anno, gene_info, by.x="ensembl_gene_id", by.y = "Name", 
                  all.x = FALSE, all.y = TRUE)
dim(gene_info)
gene_info[1:2,]

gene_info[gene_info$ExM_min > 2e4, c("ExM_min", "hgnc_symbol")]
gene_info[gene_info$PrS_min > 2e4, c("PrS_min", "hgnc_symbol")]

gene_info$gene_symbol = gene_info$hgnc_symbol

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))

wEns = which(gene_info$gene_symbol == "" | is.na(gene_info$gene_symbol))
gene_info$gene_symbol[wEns] = gene_info$ensembl_gene_id[wEns]

table(gene_info$gene_symbol == "")
table(is.na(gene_info$gene_symbol))
```

## Prepare for reading in DE results
```{r fig.width = 4, fig.height=3}
release_name = "2023_12_JAX_RNAseq_ExtraEmbryonic"
result_dir = file.path("results", release_name)
processed_output_dir = file.path(result_dir, "processed")

all_result_files = list.files(path=processed_output_dir, pattern=".tsv", 
                              all.files=TRUE, full.names=FALSE)
all_result_files = sort(all_result_files)

extract_score <- function(x){
  x_split = str_split(x, "_")[[1]]
  x_start = 6
  x_end = length(x_split)-1
  x_d_group = paste(x_split[x_start:(x_end-1)], collapse="_")
  x_gene = x_split[x_end]
  return(c(x_d_group, x_gene))
}

df_file_info = NULL

for (x in all_result_files){
  df_file_info = rbind(df_file_info, extract_score(x))
}

df_file_info = as.data.frame(df_file_info)
colnames(df_file_info) = c("DE_group", "gene")
df_file_info

```


## Compare the number of DE genes
```{r fig.width = 8, fig.height=9}

df_file_info$items = paste(df_file_info$DE_group, 
                           df_file_info$gene, sep="_")
df_file_info$items

fc0 = log2(1.5)
p0  = 0.05

gene_symbols = df_file_info$gene
gene_symbols

table(gene_symbols %in% gene_info$gene_symbol)

genes_w_multi_ensembl = names(which(table(gene_info$gene_symbol)>1))
genes_w_multi_ensembl

sorted_de_grps = sort(unique(df_file_info$DE_group))
  
for (d_group in sorted_de_grps){
  
  df_file_info_dg = df_file_info[which(df_file_info$DE_group==d_group), ]
  
  df = data.frame(KO = df_file_info_dg$items, 
                  gene_symbol = df_file_info_dg$gene, 
                  CE_nDE = NA, KO_nDE = NA, PTC_nDE = NA, 
                  CE_padj = NA, CE_log2FoldChange = NA, 
                  KO_padj = NA, KO_log2FoldChange = NA, 
                  PTC_padj = NA, PTC_log2FoldChange = NA)

  df$Model = str_extract(df$KO, "^[a-zA-Z0-9]+")
  
  gene_ids = gene_info$ensembl_gene_id[match(df_file_info_dg$gene, gene_info$gene_symbol)]
    
  for (k in 1:nrow(df_file_info_dg)){
      
    res = fread(file.path(processed_output_dir,
                          paste0(release_name, "_", df_file_info_dg$items[k], "_DEseq2.tsv")),
                          data.table = FALSE)
    print(res[1:2,])
    
    gene_id = gene_ids[k]
    w_gene = which(res$gene_ID == gene_id)
    stopifnot(length(w_gene) == 1)
    stopifnot(!(df$gene_symbol[k]%in%genes_w_multi_ensembl))
    
    for(ko1 in c("CE", "KO", "PTC")){
      
      fc = paste0(df$gene_symbol[k], "_", ko1, "_log2FoldChange")
      padj = paste0(df$gene_symbol[k], "_", ko1, "_padj")
      col_name = paste0(ko1, "_nDE")
      df[k,col_name] = sum(abs(res[[fc]]) > fc0 & res[[padj]] < p0, na.rm = TRUE)
      
      col_name = paste0(ko1, "_log2FoldChange")
      df[k,col_name] = res[[fc]][w_gene]
      
      col_name = paste0(ko1, "_padj")
      df[k,col_name] = res[[padj]][w_gene]
      
    }
    
  }
  
  print(df)

  p1 <- ggplot(df, aes(x=(PTC_nDE), y=(CE_nDE), 
                 label=gene_symbol)) + 
        geom_point() + scale_shape_manual(values = c(7, 16)) + 
        geom_text_repel() + ggtitle("# of DE genes") + 
        labs(x="PTC", y="CE") + 
        geom_abline(intercept = 0, slope = 1, color = "orange", linetype = "dashed")

  p2 <- ggplot(df, aes(x=(PTC_nDE), y=(KO_nDE),  
                 label=gene_symbol)) + 
        geom_point() + scale_shape_manual(values = c(7, 16)) + 
        geom_text_repel() + ggtitle("# of DE genes") + 
        labs(x="PTC", y="KO") + 
        geom_abline(intercept = 0, slope = 1, color = "orange", linetype = "dashed")
  
  p3 <- ggplot(df, aes(x=CE_log2FoldChange, y=-log10(CE_padj), 
                 label=gene_symbol)) + 
        geom_point(size=2) + scale_shape_manual(values = c(7, 16)) + 
        geom_text_repel() + ggtitle("DE results of the KO genes, CE") + 
        labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
        geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
        geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
                   linetype = "dashed")
  
  p4 <- ggplot(df, aes(x=KO_log2FoldChange, y=-log10(KO_padj), 
                 label=gene_symbol)) + 
        geom_point(size=2) + scale_shape_manual(values = c(7, 16)) + 
        geom_text_repel() + ggtitle("DE results of the KO genes, KO") + 
        labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
        geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
        geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
                   linetype = "dashed")
  
  p5 <- ggplot(df, aes(x=PTC_log2FoldChange, y=-log10(PTC_padj), 
                 label=gene_symbol)) + 
        geom_point(size=2) + scale_shape_manual(values = c(7, 16)) + 
        geom_text_repel() + ggtitle("DE results of the KO genes, PTC") + 
        labs(x="log2 fold change", y="-log10(adjusted p-value)")  + 
        geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
        geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
                   linetype = "dashed")

 g_combined <- ggarrange(p1, p2, p3, p4, p5, ncol = 2, nrow = 3)
 print(annotate_figure(g_combined,
                      top = text_grob(d_group, 
                                      face = "bold", size = 16)))
  
  
}


```


## Generate volcano plots

Also output the DE gene sets for the gene set enrichment analysis. 

```{r fig.width = 4, fig.height=3}

df_file_info

fc0
p0

for(k in 1:nrow(df_file_info)){
  
  it_k = df_file_info$items[k]
  d_group = df_file_info$DE_group[k]
  gene_name = df_file_info$gene[k]
  
  # create a list to store DE gene set for the gene set enrichment analysis
  de_gene_sets = list() 
  
  print(it_k)
  
  res = fread(file.path(processed_output_dir,
                        paste0(release_name, "_", it_k, "_DEseq2.tsv")),
              data.table = FALSE)

  for (ko1 in c("CE", "KO", "PTC")){

    res_k = res[,c(1, grep(ko1, names(res)))]
    names(res_k) = gsub(paste0(gene_name, "_", ko1, "_"), "", names(res_k))

    ww_up = which((res_k$log2FoldChange > fc0) & (res_k$padj < p0))
    ww_down = which((res_k$log2FoldChange < ((-1)*fc0)) & (res_k$padj < p0))

    de_gene_sets[[paste0(gene_name, "_", ko1, "_up")]] = res_k[ww_up,]$gene_ID
    de_gene_sets[[paste0(gene_name, "_", ko1, "_down")]] = res_k[ww_down,]$gene_ID
    
    res_k$DE = "No"
    res_k$DE[ww_up] <- "Up"
    res_k$DE[ww_down] <- "Down"
    
    res_k$log2FoldChange[which(res_k$log2FoldChange > 3)] = 3
    res_k$log2FoldChange[which(res_k$log2FoldChange < -3)] = -3
  
    col2use = c("blue", "grey", "red")
    names(col2use) = c("Down", "No", "Up")
    
    res_k$delabel = NA
    res_k$gene_symbol = gene_info$gene_symbol[match(res_k$gene_ID, 
                                                    gene_info$ensembl_gene_id)]
    w_de = which(res_k$DE != "No")
    res_k$delabel[w_de] = res_k$gene_symbol[w_de]
  
    print(table(res_k$DE))
    
    g2 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                      col=DE, label=delabel)) + 
           geom_point() + ggtitle(paste0(d_group, "\n", gene_name, "_", ko1)) + 
           geom_text_repel(point.padding = 0.5,
                           min.segment.length = 0, 
                           size = 3,  # Adjust text size
                           box.padding = 0.5,
                           max.overlaps = 20, 
                           show.legend = FALSE) + 
           scale_color_manual(values=col2use)
      
    print(g2)

  }
  
  # save out the DE gene lists for gene set enrichment analysis
  DE_set_output_dir = sprintf("%s/DE_gene_lists", result_dir)
                       
    
  if (!file.exists(DE_set_output_dir)){
    dir.create(DE_set_output_dir)
  }
  
  saveRDS(de_gene_sets, 
          sprintf("%s/%s_%s_DE_gene_list.rds", 
                  DE_set_output_dir, release_name, it_k))
  
}


```
      
## Check the results of one KO

For KO of FOSB, check those with significant results by PTC, and insignificant results by KO.

```{r fig.width = 5, fig.height=3}
meta$run_id_short = str_replace(meta$run_id, "^[^-]*-", "")

gene = "FOSB"

res = fread(file.path(processed_output_dir,
                      paste0(release_name, "_PrS_nor_FOSB_DEseq2.tsv")),
                      data.table = FALSE)
    
targets = which(res$FOSB_KO_padj > 0.01 & 
                  res$FOSB_PTC_padj < 1e-10)
length(targets)
res$gene_ID[targets]

w_sample = which(meta$`ALTERED GENE SYMBOLS` %in% c(gene, NA))
meta_m = meta[w_sample,]
data_m = cts_dat[,w_sample]
q75 = apply(data_m, 2, quantile, probs=0.75)
r1 = res[res$gene_ID =="ENSG00000163870", grep("FOSB", names(res))]
t(r1)

meta_m$y = data_m[which(rownames(data_m) == "ENSG00000163870"),]/q75
table(meta_m$ko_strategy)

meta_m$ko_strategy = factor(meta_m$ko_strategy)
ggplot(meta_m, aes(x = ko_strategy, y = y, color = run_id_short)) + 
  geom_boxplot(outlier.shape = NA) +  ggtitle("FOSB -> TRPA1") + 
    geom_jitter(width = 0.2, alpha = 0.5)
```

```{r}
gc()
sessionInfo()
```

