---
title: "Differential expression analysis for 2024-05 release"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Differential expression using the bulk RNAseq data from the May release.

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

path = "~/research/data/MorPhiC/May-2024/JAX_RNAseq2_ExtraEmbryonic/"
dat_path = paste0(path, "processed/release110-gencode44/Tables")
```


## Read in meta data
```{r fig.width = 3, fig.height=3}
meta = fread("data/JAX_RNAseq2_ExtraEmbryonic_meta_data.tsv", 
             data.table = FALSE)
dim(meta)
meta[1:2,]
names(meta)

table(table(meta$biomaterial_id))
table(table(meta$lib_biomaterial_id))
table(table(meta$differentiated_biomaterial_id))
```


## Read in gene count data and filter genes

```{r fig.width = 4, fig.height=3}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

stopifnot(setequal(names(cts)[-1], meta$file_id))

meta = meta[match(names(cts)[-1], meta$file_id),]
table(names(cts)[-1] == meta$file_id)

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}

model_s = meta$model_system
get_q75 <- function(x){quantile(x, probs = 0.75)}

gene_info = data.frame(Name = cts$Name, 
                       t(apply(cts_dat, 1, tapply, model_s, min)), 
                       t(apply(cts_dat, 1, tapply, model_s, median)), 
                       t(apply(cts_dat, 1, tapply, model_s, get_q75)), 
                       t(apply(cts_dat, 1, tapply, model_s, max)))

names(gene_info)[2:9] = paste0(rep(c("ExM_", "PrS_"), times=4), 
                               rep(c("min", "median", "q75", "max"), each=2))
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$ExM_q75 > 0, gene_info$PrS_q75 > 0)

w2kp = which(gene_info$ExM_q75 > 0 | gene_info$PrS_q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]

meta$read_depth = colSums(cts_dat)/1e6
meta$q75 = apply(cts_dat, 2, quantile, probs = 0.75)

ggplot(meta, aes(x=read_depth, y=q75, color=ko_strategy, shape = model_system)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  scale_shape_manual(values = c(7, 16)) + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "KO type", shape = "Model") + 
  scale_color_brewer(palette = "Set1")

meta$run_id_short = str_replace(meta$run_id, "^[^-]*-", "")
ggplot(meta, aes(x=read_depth, y=q75, color=run_id_short, shape = model_system)) +
    geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  scale_shape_manual(values = c(7, 16)) + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "Run ID", shape = "Model") + 
  scale_color_brewer(palette = "Set1")
```

# Check the effect of run id among WT

```{r fig.width = 4, fig.height=3}
sample2kp = which(meta$ko_gene == "WT")
cts_dat_m = cts_dat[,sample2kp]
meta_m    = meta[sample2kp,]
summary(meta_m$q75)
dim(cts_dat_m)

stopifnot(all(meta_m$file_id == names(cts_dat_m)))

q75 = apply(cts_dat_m, 1, quantile, probs=0.75)
cts_dat_n = t(cts_dat_m[q75 > 0,])
cts_dat_n = log(median(meta_m$q75)*cts_dat_n/meta_m$q75 + 1)
dim(cts_dat_n)

sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2
eigen_vals[1:5]/sum(eigen_vals)
pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
pvs

PC_df = data.frame(pc_scores)
PC_df = cbind(PC_df, meta_m)

gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=model_system, 
                        color=run_id_short)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
    scale_shape_manual(values = c(7, 15, 16, 17)) + 
  labs(color="KO gene", shape ="KO strategy") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("Wild type samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)
```


## DE analysis with respect to model system for WT samples

```{r fig.width = 4, fig.height=3, message=FALSE, warning=FALSE}
## Generate sample information matrix

cols2kp = c("model_system", "q75")
sample_info = meta_m[,cols2kp]

dim(sample_info)
sample_info[1:2,]

sample_info$log_q75 = log(sample_info$q75)

dds = DESeqDataSetFromMatrix(cts_dat_m, colData=sample_info, 
                                  ~ model_system + log_q75)
dds = DESeq(dds)

## association with read-depth
res_rd = results(dds, name = "log_q75")
res_rd = as.data.frame(res_rd)
dim(res_rd)
res_rd[1:2,]

table(is.na(res_rd$padj))

g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
  geom_histogram(color="darkblue", fill="lightblue", 
                 breaks = seq(0,1,by=0.01)) + 
  ggtitle("Read depth")
print(g0)

## association with model_system
dds_lrt = DESeq(dds, test="LRT", reduced = ~ + log_q75)
res_lrt = results(dds_lrt)

res_model_system = as.data.frame(res_lrt)
dim(res_model_system)
res_model_system[1:2,]

table(is.na(res_model_system$padj))

g0 = ggplot(res_model_system %>% subset(!is.na(padj)), aes(x=pvalue)) + 
  geom_histogram(color="darkblue", fill="lightblue", 
                 breaks = seq(0,1,by=0.01)) + 
  ggtitle("Model system")
print(g0)
```

# Analyze samples of different model system

```{r fig.width = 4.5, fig.height=3, message=FALSE, warning=FALSE}
table(meta$run_id, meta$model_system)
table(meta$run_id, meta$ko_gene)
table(meta$run_id, meta$model_system, meta$ko_gene)
table(meta$ko_strategy, meta$ko_gene)

for(model1 in unique(meta$model_system)){
  print(model1)
  sample2kp = which(meta$model_system == model1)
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]

  table(meta_m$ko_strategy, substr(colnames(cts_dat_m), 1, 7))
  stopifnot(all(meta_m$file_id == names(cts_dat_m)))
  q75 = apply(cts_dat_m, 1, quantile, probs=0.75)

  cts_dat_n = t(cts_dat_m[q75 > 0,])
  cts_dat_n = log(median(meta_m$q75)*cts_dat_n/meta_m$q75 + 1)
  dim(cts_dat_n)
  
  sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
  names(sample_pca)
  pc_scores = sample_pca$x
  eigen_vals = sample_pca$sdev^2
  eigen_vals[1:5]/sum(eigen_vals)
  pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
  pvs
  
  PC_df = data.frame(pc_scores)
  PC_df = cbind(PC_df, meta_m)
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=ko_strategy, color=ko_gene)) +
    geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
      scale_shape_manual(values = c(7, 15, 16, 17)) + 
    xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
    ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
    ggtitle(model1) + 
    guides(
      color = guide_legend(title = NULL, order = 1),
      shape = guide_legend(title = NULL, order = 2)
    ) +
    theme(
      legend.margin = margin(0, 0, 0, 0), 
      legend.box.just = "left", 
      legend.direction = "vertical" 
    )

  print(gPC)
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=run_id_short, color=ko_gene)) +
    geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
      scale_shape_manual(values = c(7, 15, 16, 17)) + 
    xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
    ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
    ggtitle(model1) + 
    guides(
      color = guide_legend(title = NULL, order = 1),
      shape = guide_legend(title = NULL, order = 2)
    ) +
    theme(
      legend.margin = margin(0, 0, 0, 0), 
      legend.box.just = "left", 
      legend.direction = "vertical" 
    )

  print(gPC)
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=run_id_short, color=ko_gene, 
                          alpha = ifelse(ko_gene == "WT", 1, 0.8))) +
    geom_point(size=2.5) + 
    scale_color_brewer(palette = "Dark2") + 
    scale_shape_manual(values = c(7, 15, 16, 17)) + 
    xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
    ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
    ggtitle(model1) + guides(alpha = "none") + 
    guides(
      color = guide_legend(title = NULL, order = 1),
      shape = guide_legend(title = NULL, order = 2)
    ) +
    theme(
      legend.margin = margin(0, 0, 0, 0), 
      legend.box.just = "left", 
      legend.direction = "vertical" 
    )

  print(gPC)
  
  table(meta_m$run_id, meta_m$ko_gene)
  
  ## Generate sample information matrix
  
  cols2kp = c("ko_strategy", "ko_gene", "run_id", "q75")
  sample_info = meta_m[,cols2kp]
  
  dim(sample_info)
  sample_info[1:2,]
  
  table(sample_info$ko_strategy, sample_info$ko_gene)
  
  sample_info$ko_group = paste(sample_info$ko_gene, 
                               sample_info$ko_strategy, sep="_")
  table(sample_info$ko_group)

  sample_info$log_q75 = log(sample_info$q75)
  
  dds = DESeqDataSetFromMatrix(cts_dat_m, colData=sample_info, 
                                  ~ ko_group + run_id + log_q75)

  dds = DESeq(dds)
  
  ## association with read-depth
  res_rd = results(dds, name = "log_q75")
  res_rd = as.data.frame(res_rd)
  dim(res_rd)
  res_rd[1:2,]
  
  table(is.na(res_rd$padj))
  
  g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
    geom_histogram(color="darkblue", fill="lightblue", 
                   breaks = seq(0,1,by=0.01)) + 
    ggtitle(paste0(model1, " read depth"))
  print(g0)
  
  ## association with run_id
  dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_group + log_q75)
  res_lrt = results(dds_lrt)

  res_run_id = as.data.frame(res_lrt)
  dim(res_run_id)
  res_run_id[1:2,]

  table(is.na(res_run_id$padj))
  
  g0 = ggplot(res_run_id %>% subset(!is.na(padj)), aes(x=pvalue)) + 
    geom_histogram(color="darkblue", fill="lightblue", 
                   breaks = seq(0,1,by=0.01)) + 
    ggtitle(paste0(model1, " Run ID"))
  print(g0)
  
  ## DE analysis for each KO gene
  table(sample_info$ko_gene)
  
  genos = setdiff(unique(sample_info$ko_gene), "WT")
  genos
  
  ko_grp  = c("CE", "KO", "PTC")
  res_geno_df = NULL
    
  for(geno in genos){
    
    res_geno = list()
    
    for(k_grp1 in ko_grp){
      res_g1 = results(dds, contrast = c("ko_group", 
                                         paste0(geno, "_", k_grp1), "WT_WT"))
      res_g1 = signif(data.frame(res_g1), 3)
      res_geno[[k_grp1]] = res_g1
      
      g1 = ggplot(res_g1 %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(geno, "_", k_grp1))
      print(g1)

      tag1 = sprintf("%s_%s_%s_", model1, geno, k_grp1)
      colnames(res_g1) = paste0(tag1, colnames(res_g1))
      
      if(is.null(res_geno_df)){
        res_geno_df = res_g1
      }else if(!is.null(res_geno_df)){
        stopifnot(all(rownames(res_geno_df) == rownames(res_g1)))
        res_geno_df = cbind(res_geno_df, res_g1)
      }
    }
    
    get_index <- function(x){
      which(x$padj < 0.05 & abs(x$log2FoldChange) > log2(1.5))
    }
    
    w_ce = get_index(res_geno[["CE"]])
    w_ko = get_index(res_geno[["KO"]])
    w_ptc = get_index(res_geno[["PTC"]])

    m = make_comb_mat(list("CE" = rownames(res_geno[["CE"]])[w_ce], 
              "KO" = rownames(res_geno[["KO"]])[w_ko], 
              "PTC" = rownames(res_geno[["PTC"]])[w_ptc]))
    g_up = UpSet(m)
    
    print(g_up)
    
    df1 = data.frame(padj_CE = res_geno[["CE"]]$padj, 
                     padj_KO = res_geno[["KO"]]$padj, 
                     padj_PTC = res_geno[["PTC"]]$padj)

    cr1 = cor(df1$padj_PTC, df1$padj_CE, method = "spearman", use="pair")
    cr2 = cor(df1$padj_PTC, df1$padj_KO, method = "spearman", use="pair")
    
    g4 = ggplot(data = df1, mapping = aes(x = -log10(padj_PTC), 
                                          y = -log10(padj_CE))) +
      geom_pointdensity() + ggtitle(sprintf("%s, Spearman r = %.2f", geno, cr1)) + 
      scale_color_viridis()
    
    g5 = ggplot(data = df1, mapping = aes(x = -log10(padj_PTC), 
                                          y = -log10(padj_KO))) +
      geom_pointdensity() + ggtitle(sprintf("%s, Spearman r = %.2f", geno, cr2)) + 
      scale_color_viridis()
    
    print(g4)
    print(g5)
  }
  
  dim(res_geno_df)
  res_geno_df[1:2,]
  
  res_df = data.frame(gene_ID=rownames(res_geno_df), res_geno_df)
  dim(res_df)
  res_df[1:2,]

  fwrite(res_df, sprintf("results/2024_05_%s_DEseq2.tsv", model1), sep="\t")
}
```

```{r}
gc()
sessionInfo()
```

