---
title: "Check the results of differential expression analysis"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
theme_set(theme_classic())

dat_path = "~/research/data/MorPhiC/December-2023/JAX_RNAseq_ExtraEmbryonic/"
dat_path = paste0(dat_path, "processed/Tables")
```


## Read in meta data
```{r fig.width = 3, fig.height=3}
meta_flat = read_excel("data/JAX_RNAseq_ExtraEmbryonic_meta_falttened.xlsx")
meta_flat = as.data.frame(meta_flat)
dim(meta_flat)
meta_flat[1:2,]
names(meta_flat)
```

### Gather cell line id information
```{r fig.width = 3, fig.height=3}
meta_flat$`DIFFERENTIATED CELL LINE ID`[1:5]
cell_line_id = str_split(meta_flat$`DIFFERENTIATED CELL LINE ID`, pattern = '-')
table(sapply(cell_line_id, length))

cell_line_id = lapply(cell_line_id, function(x) { length(x) <- 4; x})
cell_line_id = as.data.frame(do.call(rbind, cell_line_id))
dim(cell_line_id)
names(cell_line_id) = c("model_system", "cell_line", "clone_id", "condition")
cell_line_id[1:2,]
cell_line_id$condition[which(is.na(cell_line_id$condition))] = "nor"

table(meta_flat$`Model System`, cell_line_id$model_system)
table(cell_line_id$model_system, cell_line_id$condition, useNA="ifany")
```

### Add ```run_id``` and ```ko_strategy```. 

```{r fig.width = 3, fig.height=3}
w2update = which(meta_flat$`RUN ID` == "20231004_23-robson-008-run2")
meta_flat$run_id = meta_flat$`RUN ID`
meta_flat$run_id[w2update] = "20230918_23-robson-008"

table(meta_flat$run_id, cell_line_id$model_system, useNA="ifany")
table(meta_flat$run_id, cell_line_id$condition, useNA="ifany")

table(meta_flat$`TARGETED GENOMIC REGION`, useNA="ifany")
meta_flat$ko_strategy = str_extract(meta_flat$`TARGETED GENOMIC REGION`, 
                                    "\\(\\S+\\)$")
meta_flat$ko_strategy = str_remove_all(meta_flat$ko_strategy, "\\(|\\)")
meta_flat$ko_strategy[which(is.na(meta_flat$ko_strategy))] = "WT"

table(meta_flat$`TARGETED GENOMIC REGION`, meta_flat$ko_strategy, 
      useNA="ifany")

meta_flat$ko_gene = meta_flat$`ALTERED GENE SYMBOLS`
meta_flat$ko_gene[which(is.na(meta_flat$ko_gene))] = "WT"

table(cell_line_id$model_system, meta_flat$ko_gene)
table(meta_flat$run_id, meta_flat$ko_gene)
table(meta_flat$ko_strategy, meta_flat$ko_gene)
```

## Read in gene count data and filter genes

```{r fig.width = 4, fig.height=3}
meta = cbind(meta_flat, cell_line_id)
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]

setequal(names(cts)[-1], meta$`Counts file`)

meta = meta[match(names(cts)[-1], meta$`Counts file`),]
table(names(cts)[-1] == meta$`Counts file`)

cts_dat = data.matrix(cts[,-1])
rownames(cts_dat) = cts$Name
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r fig.width = 4, fig.height=3}
model_s = meta$model_system
get_q75 <- function(x){quantile(x, probs = 0.75)}

gene_info = data.frame(Name = cts$Name, 
                       t(apply(cts_dat, 1, tapply, model_s, min)), 
                       t(apply(cts_dat, 1, tapply, model_s, get_q75)))

names(gene_info)[2:5] = c("ExM_min", "PrS_min", "ExM_q75", "PrS_q75")
dim(gene_info)
gene_info[1:2,]

summary(gene_info[,-1])

table(gene_info$ExM_q75 > 0, gene_info$PrS_q75 > 0)

w2kp = which(gene_info$ExM_q75 > 0 | gene_info$PrS_q75 > 0)
cts_dat = cts_dat[w2kp,]
dim(cts_dat)

gene_info = gene_info[w2kp,]
```

## Read in gene annoation
```{r fig.width = 4, fig.height=3}
gene_anno = fread("data/gencode_v44_basic_gene_info.tsv")
dim(gene_anno)
gene_anno[1:2,]

table(gene_info$Name %in% gene_anno$ensembl_gene_id)
# the difference are genes not on chr1-22 or chrM
setdiff(gene_info$Name, gene_anno$ensembl_gene_id)

gene_info = merge(gene_anno, gene_info, by.x="ensembl_gene_id", by.y = "Name", 
                  all.x = FALSE, all.y = TRUE)
dim(gene_info)
gene_info[1:2,]

gene_info[gene_info$ExM_min > 5e4, c("ExM_min", "hgnc_symbol")]

gene_info$gene_symbol = gene_info$hgnc_symbol

wEns = which(gene_info$gene_symbol == "" | is.na(gene_info$gene_symbol))
gene_info$gene_symbol[wEns] = gene_info$ensembl_gene_id[wEns]
```

## Read in DE results
```{r fig.width = 4, fig.height=3}
PrS_nor = fread("results/2023_12_PrS_nor_DEseq2.tsv", data.table = FALSE)
dim(PrS_nor)
PrS_nor[1:2,1:7]

PrS_hyp = fread("results/2023_12_PrS_hyp_DEseq2.tsv", data.table = FALSE)
dim(PrS_hyp)
PrS_hyp[1:2,1:7]

ExM_nor = fread("results/2023_12_ExM_nor_DEseq2.tsv", data.table = FALSE)
dim(ExM_nor)
ExM_nor[1:2,1:7]

res = merge(PrS_nor, PrS_hyp)
res = merge(res, ExM_nor)

dim(res)
res[1:2,1:5]
```

## Compare the number of DE genes
```{r fig.width = 4, fig.height=3}
items = names(res)[grep("_padj", names(res))]
items = unique(str_extract(items, "^[a-zA-Z0-9]+_[a-zA-Z0-9]+_[a-zA-Z0-9]+"))

fc0 = log2(1.5)
p0  = 0.01

gene_symbols = str_extract(items, "[a-zA-Z0-9]+$")
gene_symbols
table(gene_symbols %in% gene_info$gene_symbol)

df = data.frame(KO = items, gene_symbol = gene_symbols, 
                CE_nDE = NA, KO_nDE = NA, PTC_nDE = NA, 
                CE_padj = NA, CE_log2FoldChange = NA, 
                KO_padj = NA, KO_log2FoldChange = NA, 
                PTC_padj = NA, PTC_log2FoldChange = NA)

df$Model = str_extract(df$KO, "^[a-zA-Z0-9]+")
df$Condition = gsub("_", "", str_extract(df$KO, "_([a-zA-Z0-9]+)_"))

gene_ids = gene_info$ensembl_gene_id[match(gene_symbols, gene_info$gene_symbol)]

for(k in 1:length(items)){
  it_k = items[k]
  gene_id = gene_ids[k]
  w_gene = which(res$gene_ID == gene_id)
  stopifnot(length(w_gene) == 1)
  
  for(ko1 in c("CE", "KO", "PTC")){
    fc = paste0(it_k, "_", ko1, "_log2FoldChange")
    padj = paste0(it_k, "_", ko1, "_padj")
    col_name = paste0(ko1, "_nDE")
    df[k, col_name] = sum(abs(res[[fc]]) > fc0 & res[[padj]] < p0, na.rm = TRUE)
    
    col_name = paste0(ko1, "_log2FoldChange")
    df[k,col_name] = res[[fc]][w_gene]
    
    col_name = paste0(ko1, "_padj")
    df[k,col_name] = res[[padj]][w_gene]
  }
}

df

ggplot(df, aes(x=(PTC_nDE), y=(CE_nDE), col=Model, shape=Condition, 
               label=gene_symbol)) + 
    geom_point() +   scale_shape_manual(values = c(7, 16)) + 
    geom_text_repel() + ggtitle("# of DE genes") + 
    labs(x="PTC", y="CE") + 
    geom_abline(intercept = 0, slope = 1, color = "orange", linetype = "dashed")

ggplot(df, aes(x=(PTC_nDE), y=(KO_nDE), col=Model, shape=Condition, 
               label=gene_symbol)) + 
    geom_point() +   scale_shape_manual(values = c(7, 16)) + 
    geom_text_repel() + ggtitle("# of DE genes") + 
    labs(x="PTC", y="KO") + 
    geom_abline(intercept = 0, slope = 1, color = "orange", linetype = "dashed")

ggplot(df, aes(x=CE_log2FoldChange, y=-log10(CE_padj), col=Model, 
               shape=Condition, label=gene_symbol)) + 
    geom_point() +   scale_shape_manual(values = c(7, 16)) + 
    geom_text_repel() + ggtitle("CE") + 
    labs(x="log2 fold change", y="-log10(p-value)")  + 
    geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
    geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
               linetype = "dashed")

ggplot(df, aes(x=KO_log2FoldChange, y=-log10(KO_padj), col=Model, 
               shape=Condition, label=gene_symbol)) + 
    geom_point() +   scale_shape_manual(values = c(7, 16)) + 
    geom_text_repel() + ggtitle("KO") + 
    labs(x="log2 fold change", y="-log10(p-value)")  + 
    geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
    geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
               linetype = "dashed")

ggplot(df, aes(x=PTC_log2FoldChange, y=-log10(PTC_padj), 
               col=Model, shape=Condition, label=gene_symbol)) + 
    geom_point() + scale_shape_manual(values = c(7, 16)) + 
    geom_text_repel() + ggtitle("PTC") + 
    labs(x="log2 fold change", y="-log10(p-value)")  + 
    geom_hline(yintercept = 2, color = "grey", linetype = "dashed") + 
    geom_vline(xintercept = c(log2(1/1.5), log2(1.5)), color = "grey",
               linetype = "dashed")

```

## Generate valcano plots
```{r fig.width = 4, fig.height=3}
items = names(res)[grep("_padj", names(res))]
items

for(it1 in items){
  id1 = gsub("_padj", "", it1)
  
  res_k = res[,c(1, grep(id1, names(res)))]
  names(res_k) = gsub(paste0(id1, "_"), "", names(res_k))
  
  res_k$DE = "No"
  res_k$DE[res_k$log2FoldChange > log2(1.5) & res_k$padj < 0.01] <- "Up"
  res_k$DE[res_k$log2FoldChange < -log2(1.5) & res_k$padj < 0.01] <- "Down"
  
  res_k$log2FoldChange[which(res_k$log2FoldChange > 3)] = 3
  res_k$log2FoldChange[which(res_k$log2FoldChange < -3)] = -3

  col2use = c("blue", "grey", "red")
  names(col2use) = c("Down", "No", "Up")
  
  res_k$delabel = NA
  res_k$gene_symbol = gene_info$gene_symbol[match(res_k$gene_ID, 
                                                  gene_info$ensembl_gene_id)]
  w_de = which(res_k$DE != "No")
  res_k$delabel[w_de] = res_k$gene_symbol[w_de]

  table(res_k$DE)
  
  g2 = ggplot(res_k, aes(x=log2FoldChange, y=-log10(padj), 
                  col=DE, label=delabel)) + 
    geom_point() + ggtitle(id1) + 
    geom_text_repel() + 
    scale_color_manual(values=col2use)
  
  print(g2)
}

fwrite(res, file = "results/2023_12_DEseq2.tsv", append = FALSE, sep = "\t")
system("gzip results/2023_12_DEseq2.tsv")
```
      
## Check the results of one KO

For KO of FOSB, check those with significant results by PTC, and insignificant results by KO.

```{r fig.width = 5, fig.height=3}
meta$run_id_short = str_replace(meta$run_id, "^[^-]*-", "")

gene = "FOSB"

targets = which(res$PrS_nor_FOSB_KO_padj > 0.01 & 
                  res$PrS_nor_FOSB_PTC_padj < 1e-10)
length(targets)
res$gene_ID[targets]

w_sample = which(meta$`ALTERED GENE SYMBOLS` %in% c(gene, NA))
meta_m = meta[w_sample,]
data_m = cts_dat[,w_sample]
q75 = apply(data_m, 2, quantile, probs=0.75)
r1 = res[res$gene_ID =="ENSG00000163870", grep("FOSB", names(res))]
t(r1)

meta_m$y = data_m[which(rownames(data_m) == "ENSG00000163870"),]/q75
table(meta_m$ko_strategy)

meta_m$ko_strategy = factor(meta_m$ko_strategy)
ggplot(meta_m, aes(x = ko_strategy, y = y, color = run_id_short)) + 
  geom_boxplot(outlier.shape = NA) +  ggtitle("FOSB -> TRPA1") + 
    geom_jitter(width = 0.2, alpha = 0.5)
```

```{r}
gc()
sessionInfo()
```

