---
title: "Differential expression analysis for 2025-06 release JAX_RNAseq11"
author: "Bo Yu, Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# explicitly set warning level in R, to prevent a warning from being converted to an error
options(warn=1)
```

Differential expression using the bulk RNAseq JAX_RNAseq11 data from the June 2025 release.

All samples are WT and from same model system, condition. Only draw PCA plots.

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)
library(viridis)
library(ggpointdensity)
library(tidyverse)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
library(ggsci)
library(ggprism)
theme_set(theme_classic())
```

## Load data of two releases & merge

```{r}
# Load meta data
meta1 = fread("data/June_2025/JAX_RNAseq11_meta_data.tsv", data.table = FALSE)
table(meta1$timepoint_value, meta1$run_id, meta1$model_system, meta1$treatment_condition, meta1$ko_gene)
meta1$timepoint_value = as.factor(meta1$timepoint_value)
meta1$condition = 'nor'
# Load count data
path1 = "../../MorPhiC/data/MorPhiC_internal_releases/MorPhiC_Release_June_2025/JAX_RNAseq11/"
dat_path1 = file.path(path1, "processed/Tables")
count1 = fread(file.path(dat_path1, "genesCounts.csv"), data.table = FALSE) |>
  column_to_rownames("Name") |>
  as.matrix() |>
  t()
# Align samples
stopifnot(
  all(!duplicated(meta1$file_id)),
  all(!duplicated(rownames(count1))),
  setequal(meta1$file_id, rownames(count1))
)
count1 = count1[meta1$file_id, ]
```

```{r}
# Load 2024-06 release JAX_RNAseq_Neuroectoderm data
# Load meta data
meta2 = fread("data/June_2024/JAX_RNAseq_Neuroectoderm_meta_data.tsv", data.table = FALSE)
table(meta2$timepoint_value, meta2$run_id, meta2$model_system, meta2$treatment_condition)
meta2$timepoint_value = as.factor(meta2$timepoint_value)
meta2$condition = 'nor'
# Load count data
path2 = "../../MorPhiC/data/June-2024/JAX_RNAseq_Neuroectoderm/"
dat_path2 = paste0(path2, "processed/release110-gencode44/Tables")
count2 = fread(file.path(dat_path2, "genesCounts.csv"), data.table = FALSE) |>
  column_to_rownames("Name") |>
  as.matrix() |>
  t()
# Align samples
stopifnot(
  all(!duplicated(meta2$file_id)),
  all(!duplicated(rownames(count2))),
  setequal(meta2$file_id, rownames(count2))
)
count2 = count2[meta2$file_id, ]
```

```{r}
# Merge
# Align metadata from different releases
stopifnot(identical(names(meta1), names(meta2)))
meta = rbind(
  tibble(release = '2025-06', meta1),
  tibble(release = '2024-06', meta2)
)
stopifnot(!any(duplicated(meta$file_id)))
table(meta$release, meta$run_id)
sort(unique(meta$timepoint_value))
timepoint_values = sort(as.integer(as.character(unique(meta$timepoint_value))))
timepoint_values
meta$timepoint_value = factor(meta$timepoint_value, levels = timepoint_values)
# Align counts from different releases
stopifnot(identical(colnames(count1), colnames(count2)))
count = rbind(count1, count2)
```

## PCA shows strong batch effect of new run_id

```{r, fig.width=5, fig.height=4}
meta$q75 = apply(count, 1, quantile, probs = 0.75)
summary(meta$q75)

run_pca <- \(count, meta) {
  q75 = apply(count, 2, quantile, probs=0.75)
  cts_dat_n = count[, q75 > 0]
  cts_dat_n = log(median(meta$q75)*cts_dat_n/meta$q75 + 1)
  
  sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
  pc_scores = sample_pca$x
  eigen_vals = sample_pca$sdev^2
  eigen_vals[1:5]/sum(eigen_vals)
  pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
  
  PC_df = data.frame(pc_scores)
  PC_df = cbind(PC_df, meta)
  list(PC_df = PC_df, pvs = pvs)
}

res_pca = run_pca(count, meta)
PC_df = res_pca$PC_df
pvs = res_pca$pvs

gPC = ggplot(PC_df, aes(x = PC1, y = PC2, color=timepoint_value, shape = release)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Set3") +
  scale_shape_prism() +
  labs(color="Time point") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("All samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)

gPC = ggplot(PC_df, aes(x = PC1, y = PC2, color=ko_gene, shape = release)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Set3") +
  labs(color="Time point") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("All samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)

gPC = ggplot(PC_df |> filter(ko_gene == 'WT'), aes(x = PC1, y = PC2, color=timepoint_value, shape = run_id)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Set3") +
  scale_shape_manual(values = c(0:9)) +
  labs(color="Time point") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("WT samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)
```

## Explore batch difference across day 3-5

```{r, fig.width=5, fig.height=4}
# Subset samples: WT, day 3-5
submeta = meta |> filter(timepoint_value %in% 3:5, ko_gene == 'WT')
nrow(submeta)
subcount = count[submeta$file_id, ]
# PCA plot
res_pca = run_pca(subcount, submeta)
PC_df = res_pca$PC_df
pvs = res_pca$pvs
ggplot(PC_df, aes(x = PC1, y = PC2, color=timepoint_value, shape = run_id)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
  scale_shape_prism() +
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("All samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )
```

## Define groups for DE testing

```{r}
# Subset data by matching time points
table(meta1$timepoint_value)
table(meta2$timepoint_value)
meta = meta |> 
  filter(timepoint_value %in% 3:5) |>
  droplevels()
table(meta$timepoint_value)
count = count[meta$file_id, ]
```

```{r}
# Define DE groups
table(meta$ko_gene, meta$run_id, meta$timepoint_value, meta$model_system)
table(meta$ko_gene, meta$ko_strategy, meta$timepoint_value)
```

For day 5, test FEZF2 effect under 3 KO strategies separately using count ~ run_id + ko.

For day 3 and 4, combine their WT samples to correct for batch effect better. Test LHX5 effect under 3 KO strategies separately using count ~ run_id + ko_day3 + ko_day4.

### Prepare output directories for DE results

```{r message=FALSE, warning=FALSE}

release_name = "2025_06_JAX_RNAseq11"

output_dir = file.path("results", release_name)
raw_output_dir = file.path(output_dir, "raw")
processed_output_dir = file.path(output_dir, "processed")

if (!file.exists(raw_output_dir)){
  dir.create(raw_output_dir, recursive = TRUE)
}

if (!file.exists(processed_output_dir)){
  dir.create(processed_output_dir, recursive = TRUE)
}

```

## DE test for day 5 FEZF2

### Format data to match previous code

```{r}
# Prepare meta
meta_full = meta
meta = meta_full |> 
  filter(timepoint_value == 5) |>
  mutate(
    gene_group = 'FEZF2',
    run_id_short = run_id,
    timepoint = timepoint_value
  ) |>
  droplevels()

# Prepare counts
count_full = count
count = count_full[meta$file_id, ]
cts = data.frame(Name = colnames(count), t(count))
colnames(cts) = c("Name", rownames(count))
cts_dat = data.matrix(cts[, -1])
rownames(cts_dat) = cts$Name

# Read in gene annotation
gene_anno = fread("data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2, ]
# Make sure all KO genes have unique ensembl gene id
gene_anno |> 
  filter(hgnc_symbol %in% setdiff(meta$ko_gene, 'WT')) |>
  arrange(hgnc_symbol)
# Exclude genes without annotation
stopifnot(all(rownames(cts_dat) %in% gene_anno$ensembl_gene_id))
```

```{r}
# Discard genes whose expression is 0 in more than 75% of samples
gene_info = data.frame(
  apply(cts_dat, 1, tapply, meta$model_system, min), 
  apply(cts_dat, 1, tapply, meta$model_system, median), 
  apply(cts_dat, 1, tapply, meta$model_system, \(x) quantile(x, probs = 0.75)), 
  apply(cts_dat, 1, tapply, meta$model_system, max)
)
gene_info[1:2, ]
names(gene_info) = paste0(
  rep(c("CBO_"), times = 4), 
  rep(c("min", "median", "q75", "max"), each = 1)
)
summary(gene_info)
table(gene_info$CBO_q75 > 0)

# Subset genes
w2kp = which(gene_info$CBO_q75 > 0)
cts_dat = cts_dat[w2kp, ]
gene_info = gene_info[w2kp, ]
```

### DE analysis for FEZF2

```{r fig.width = 4.8, fig.height=3.2, message=FALSE, warning=FALSE}

df_size = NULL

unique_model_ss = unique(meta$model_system)
unique_model_ss = sort(unique_model_ss)

for(model1 in unique_model_ss){
  
  print("")
  print("==========================================")
  print(model1)
  print("==========================================")
  print("")
  
  sample2kp = which(meta$model_system == model1)
  
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
 
  table(meta_m$ko_strategy)
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  
  ## Generate sample information matrix
  cols2kp = c("model_system", "condition", "ko_strategy", "ko_gene", "gene_group", 
              "run_id", "run_id_short", "timepoint", "q75")
  sample_info = meta_m[,cols2kp]
  colnames(sample_info)[which(colnames(sample_info)=="timepoint")] = "day"
  
  dim(sample_info)
  sample_info[1:2,]
  
  sample_info$ko_gene_gene_group = paste(sample_info$ko_gene, 
                                         sample_info$gene_group, sep="_")
  
  print(table(sample_info$run_id, sample_info$ko_gene))
  print(table(sample_info$ko_gene, sample_info$day)) 
  print(table(sample_info$ko_gene_gene_group, sample_info$day)) 
  
  # print the table of ko_gene and run_id by (day, condition) combinations
  sample_info$day_condition = paste0(sample_info$day, "_", sample_info$condition)
  unique_day_condition_model_s = unique(sample_info$day_condition)
  unique_day_condition_model_s = as.character(sort(unique_day_condition_model_s))
  
  for (cur_d_c in as.character(unique_day_condition_model_s)){
    sample_info_by_day = sample_info[which(sample_info$day_condition==as.character(cur_d_c)), ]
    print("---------------------------")
    print(paste0("Model system: ", model1))
    print(paste0("On day_condition ", as.character(cur_d_c), ": "))
    print("Cross table between knockout gene and run id:")
    print(table(sample_info_by_day$ko_gene, sample_info_by_day$run_id_short))    
    print("Cross table between knockout gene and knockout gene group:")
    print(table(sample_info_by_day$ko_gene, sample_info_by_day$gene_group))  
  }

  sample_info$model_day = paste0(sample_info$model_system, 
                                 "_day_", 
                                 as.character(sample_info$day))
    
  sample_info$de_grp = paste0(sample_info$model_day, "_", sample_info$ko_strategy)
  
  unique_de_grps = setdiff(unique(sample_info$de_grp), grep("WT", unique(sample_info$de_grp), value=TRUE))
  sorted_de_grps = sort(unique_de_grps)
  
  for (d_group in sorted_de_grps){

    dg_index = which(sample_info$de_grp==d_group | sample_info$de_grp==paste0(paste0(strsplit(d_group, "_")[[1]][-4], collapse="_"), "_WT"))
    cts_dat_m_dg = cts_dat_m[, dg_index]
    sample_info_dg = sample_info[dg_index, ]
    
    print("")
    print("-----------------------------------")
    print(paste0("Model system: ", model1))  
    if (sum(sample_info_dg$ko_gene=="WT") < 2){
      print(sprintf("DE group %s has only %d wild type samples", 
                    d_group, sum(sample_info_dg$ko_gene=="WT")))
      print(sprintf("do not run DESeq2 for it"))
      print("-----------------------------------")  
      next
    }else{
      print(sprintf("DE group %s DESeq2 results:", d_group))
      print("-----------------------------------")  
    }    
    
    
    # do two steps of filtering
    # first, filter based on q75 of each gene
    # second, filter based on whether each gene is expressed in at least 4 cells
    
    dg_q75 = apply(cts_dat_m_dg, 1, quantile, probs=0.75)
    cts_dat_m_dg = cts_dat_m_dg[dg_q75 > 0,]
    
    print(sprintf("After filtering by gene expression q75, the number of genes reduces from %d to %d", 
                  length(dg_q75), nrow(cts_dat_m_dg)))

    n_pos = rowSums(cts_dat_m_dg>0)
    cts_dat_m_dg = cts_dat_m_dg[n_pos >= 4,]

    if (length(n_pos)==nrow(cts_dat_m_dg)){
      print("After filtering by nonzero expression in at least 4 samples, the number of genes does not change.")
    }else{
      print(sprintf("After filtering by nonzero expression in at least 4 samples, the number of genes reduces from %d to %d", 
                    length(n_pos), nrow(cts_dat_m_dg)))    
    }
    
    # update the q75 based on only genes that will be used in the DE analysis
    sample_info_dg$q75 = apply(cts_dat_m_dg, 2, quantile, probs = 0.75)
    sample_info_dg$log_q75 = log(sample_info_dg$q75)
    
    sample_info_dg$ko_gene = as.factor(sample_info_dg$ko_gene)
    
    if (length(unique(sample_info_dg$run_id))==1){
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene + log_q75)
    }else{
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene + run_id + log_q75)     
    }
  
    start.time <- Sys.time()
    dds = DESeq(dds)
    end.time <- Sys.time()

    time.taken <- end.time - start.time
    print(time.taken)
    
    ## association with read-depth
    res_rd = results(dds, name = "log_q75")
    res_rd = as.data.frame(res_rd)
    dim(res_rd)
    res_rd[1:2,]
    
    table(is.na(res_rd$padj))
  
    g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
      geom_histogram(color="darkblue", fill="lightblue", 
                     breaks = seq(0,1,by=0.01)) + 
      ggtitle(paste0(d_group, "\nread depth"))
    print(g0)
  
    ## Rerun the analysis without read-depth if it is not significant for 
    ## most genes, and the number of samples is small
    ## i.e., proportion of non-null in the distribution is smaller than 0.01
    ## (this needs to be restricted to the genes with not NA adjusted pvalue)
    ## or if smaller than 0.1 and the number of samples is not greater than 6

    pi_1_rd = max(0, mean(res_rd$pvalue[!is.na(res_rd$padj)] < 0.05) - 0.05)
    pi_1_rd = max(pi_1_rd, 0, 1 - 2*mean(res_rd$pvalue[!is.na(res_rd$padj)] > 0.5))
    
    print(sprintf("prop of non-null for rd: %.2e", pi_1_rd))
    
    if(pi_1_rd < 0.01 || (ncol(cts_dat_m_dg) <= 6 && pi_1_rd < 0.1 )){
      
      print("rerun DESeq2 without read depth")
      
      include_rd = 0
      
      if (length(unique(sample_info_dg$run_id))==1){
        dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                      ~ ko_gene)
      }else{
        dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                      ~ ko_gene + run_id)     
      }
      
      dds = DESeq(dds)
      
    }else{
      include_rd = 1
    }

    ## association with run_id
      
    if (length(unique(sample_info_dg$run_id))>1){
      if(include_rd)
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene + log_q75)
      else
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene)
      
      res_lrt = results(dds_lrt)
    
      res_run_id = as.data.frame(res_lrt)
      dim(res_run_id)
      res_run_id[1:2,]
    
      table(is.na(res_run_id$padj))
      
      g0 = ggplot(res_run_id %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, "\nRun ID"))
      print(g0)
    
    }
    
    ## DE analysis for each KO gene
    table(sample_info_dg$ko_gene)
    
    genos = setdiff(unique(sample_info_dg$ko_gene), "WT")
    genos = sort(genos)
    
    for(geno in genos){
      
      res_g1 = results(dds, contrast = c("ko_gene", geno, "WT"))
      res_g1 = signif(data.frame(res_g1), 3)

      # add a column to record the number of zeros per test
      test_index = which(sample_info_dg$ko_gene%in%c(geno, "WT"))
      
      cts_dat_m_dg_test = cts_dat_m_dg[, test_index]
      n_zero = rowSums(cts_dat_m_dg_test==0)
      res_g1$n_zeros = n_zero    

      # record the number of samples involved in the comparison
      test_ko_gene_vec = sample_info_dg$ko_gene[test_index]
      
      df_size = rbind(df_size, 
                      c(d_group, 
                        paste0(geno, "_", setdiff(sample_info_dg$ko_strategy, 'WT')), 
                        sum(test_ko_gene_vec!="WT"), 
                        sum(test_ko_gene_vec=="WT")))   
      
      # prepare a processed version of output
      res_g1_reduced = res_g1[, c("log2FoldChange", "pvalue", "padj", "n_zeros")]
      res_g1_reduced$padj[which(res_g1_reduced$n_zeros>=4)] = NA
      
      if ((sum(test_ko_gene_vec!="WT")==1)|(sum(test_ko_gene_vec=="WT")==1)){
        res_g1_reduced$padj = NA
      }
      
      res_g1_reduced = res_g1_reduced[, c("log2FoldChange", "pvalue", "padj")]      
  
      n_DE = sum(res_g1_reduced$padj < 0.05 & abs(res_g1_reduced$log2FoldChange) > log2(1.5), 
                 na.rm = TRUE)
      print(paste0(d_group, " under KO gene ", geno, ":"))
      print(sprintf("number of DE genes: %d", n_DE))

      g1 = ggplot(res_g1_reduced %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, "\n", geno))
      print(g1)

      tag1 = sprintf("%s_%s_", geno, setdiff(sample_info_dg$ko_strategy, 'WT'))
      colnames(res_g1) = paste0(tag1, colnames(res_g1))
      colnames(res_g1_reduced) = paste0(tag1, colnames(res_g1_reduced))
      
      
      res_df = data.frame(gene_ID=rownames(res_g1), 
                          res_g1)
      dim(res_df)
      res_df[1:2,]
      
      
      res_reduced_gene_anno = gene_anno[match(rownames(res_g1_reduced), 
                                              gene_anno$ensembl_gene_id), ]
      stopifnot(all(rownames(res_g1_reduced)==res_reduced_gene_anno$ensembl_gene_id))     
      
      # set gene hgnc_symbol that equal "" to NA
      hgnc_symbol_vec = res_reduced_gene_anno$hgnc_symbol
      hgnc_symbol_vec[which(hgnc_symbol_vec=="")] = NA
      
      res_reduced_df = data.frame(gene_ID=rownames(res_g1_reduced), 
                                  hgnc_symbol=hgnc_symbol_vec,
                                  chr=res_reduced_gene_anno$chr, 
                                  strand=res_reduced_gene_anno$strand, 
                                  start=res_reduced_gene_anno$start, 
                                  end=res_reduced_gene_anno$end, 
                                  res_g1_reduced)
      
      print("hgnc_symbol empty string and NA tables:")
      print(table(res_reduced_df$hgnc_symbol=="", useNA="ifany"))
      print(table(is.na(res_reduced_df$hgnc_symbol)))
        
      dim(res_reduced_df)
      res_reduced_df[1:2,]      
      
      setting_name = d_group
      
      fwrite(res_df, 
             sprintf("%s/%s_%s_%s_DEseq2_raw.tsv", 
                     raw_output_dir, release_name, setting_name, geno), 
             sep="\t")
      fwrite(res_reduced_df, 
             sprintf("%s/%s_%s_%s_DEseq2.tsv", 
                     processed_output_dir, release_name, setting_name, geno), 
             sep="\t")      

    }
    
  }
  
}

colnames(df_size) = c("DE_group", "knockout_gene_strategy", "n_ko", "n_WT")
df_size = as.data.frame(df_size)

multi_rows = which(table(df_size$knockout_gene_strategy)>1)
multi_rows

df_size[which(df_size$knockout_gene_strategy%in%names(multi_rows)),]
```

## DE test for day 3 & 4 LHX5

### Format data to match previous code

```{r}
# Prepare meta
meta = meta_full |> 
  filter(timepoint_value %in% 3:4) |>
  mutate(
    gene_group = 'LHX5',
    run_id_short = run_id,
    timepoint = timepoint_value
  ) |>
  droplevels()
table(meta$ko_gene, meta$timepoint, meta$ko_strategy)

# Prepare counts
count = count_full[meta$file_id, ]
cts = data.frame(Name = colnames(count), t(count))
colnames(cts) = c("Name", rownames(count))
cts_dat = data.matrix(cts[, -1])
rownames(cts_dat) = cts$Name

# Make sure all KO genes have unique ensembl gene id
gene_anno |> 
  filter(hgnc_symbol %in% setdiff(meta$ko_gene, 'WT')) |>
  arrange(hgnc_symbol)
# Exclude genes without annotation
stopifnot(all(rownames(cts_dat) %in% gene_anno$ensembl_gene_id))
```

```{r}
# Discard genes whose expression is 0 in more than 75% of samples
gene_info = data.frame(
  apply(cts_dat, 1, tapply, meta$model_system, min), 
  apply(cts_dat, 1, tapply, meta$model_system, median), 
  apply(cts_dat, 1, tapply, meta$model_system, \(x) quantile(x, probs = 0.75)), 
  apply(cts_dat, 1, tapply, meta$model_system, max)
)
gene_info[1:2, ]
names(gene_info) = paste0(
  rep(c("CBO_"), times = 4), 
  rep(c("min", "median", "q75", "max"), each = 1)
)
summary(gene_info)
table(gene_info$CBO_q75 > 0)

# Subset genes
w2kp = which(gene_info$CBO_q75 > 0)
cts_dat = cts_dat[w2kp, ]
gene_info = gene_info[w2kp, ]
```

### DE analysis for LHX5

```{r fig.width = 4.8, fig.height=3.2, message=FALSE, warning=FALSE}

unique_model_ss = unique(meta$model_system)
unique_model_ss = sort(unique_model_ss)

for(model1 in unique_model_ss){
  
  print("")
  print("==========================================")
  print(model1)
  print("==========================================")
  print("")
  
  sample2kp = which(meta$model_system == model1)
  
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
 
  table(meta_m$ko_strategy)
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  
  ## Generate sample information matrix
  cols2kp = c("model_system", "condition", "ko_strategy", "ko_gene", "gene_group", 
              "run_id", "run_id_short", "timepoint", "q75")
  sample_info = meta_m[,cols2kp]
  colnames(sample_info)[which(colnames(sample_info)=="timepoint")] = "day"
  
  dim(sample_info)
  sample_info[1:2,]
  
  sample_info$ko_gene_gene_group = paste(sample_info$ko_gene, 
                                         sample_info$gene_group, sep="_")
  
  print(table(sample_info$run_id, sample_info$ko_gene))
  print(table(sample_info$ko_gene, sample_info$day)) 
  print(table(sample_info$ko_gene_gene_group, sample_info$day)) 
  
  # print the table of ko_gene and run_id by (day, condition) combinations
  unique_condition_model_s = unique(sample_info$condition)
  unique_condition_model_s = as.character(sort(unique_condition_model_s))
    
  sample_info$de_grp = paste0(sample_info$model_system, "_day3_and_day4_", sample_info$ko_strategy)
  
  unique_de_grps = setdiff(unique(sample_info$de_grp), grep("WT", unique(sample_info$de_grp), value=TRUE))
  sorted_de_grps = sort(unique_de_grps)
  
  for (d_group in sorted_de_grps){

    dg_index = which(sample_info$de_grp==d_group | sample_info$de_grp==paste0(paste0(strsplit(d_group, "_")[[1]][-5], collapse="_"), "_WT"))
    cts_dat_m_dg = cts_dat_m[, dg_index]
    sample_info_dg = sample_info[dg_index, ]
    
    print("")
    print("-----------------------------------")
    print(paste0("Model system: ", model1))  
    if (sum(sample_info_dg$ko_gene=="WT") < 2){
      print(sprintf("DE group %s has only %d wild type samples", 
                    d_group, sum(sample_info_dg$ko_gene=="WT")))
      print(sprintf("do not run DESeq2 for it"))
      print("-----------------------------------")  
      next
    }else{
      print(sprintf("DE group %s DESeq2 results:", d_group))
      print("-----------------------------------")  
    }    
    
    
    # do two steps of filtering
    # first, filter based on q75 of each gene
    # second, filter based on whether each gene is expressed in at least 4 cells
    
    dg_q75 = apply(cts_dat_m_dg, 1, quantile, probs=0.75)
    cts_dat_m_dg = cts_dat_m_dg[dg_q75 > 0,]
    
    print(sprintf("After filtering by gene expression q75, the number of genes reduces from %d to %d", 
                  length(dg_q75), nrow(cts_dat_m_dg)))

    n_pos = rowSums(cts_dat_m_dg>0)
    cts_dat_m_dg = cts_dat_m_dg[n_pos >= 4,]

    if (length(n_pos)==nrow(cts_dat_m_dg)){
      print("After filtering by nonzero expression in at least 4 samples, the number of genes does not change.")
    }else{
      print(sprintf("After filtering by nonzero expression in at least 4 samples, the number of genes reduces from %d to %d", 
                    length(n_pos), nrow(cts_dat_m_dg)))    
    }
    
    # update the q75 based on only genes that will be used in the DE analysis
    sample_info_dg$q75 = apply(cts_dat_m_dg, 2, quantile, probs = 0.75)
    sample_info_dg$log_q75 = log(sample_info_dg$q75)
    
    sample_info_dg$ko_gene_day3 = as.factor(sample_info_dg$ko_gene == "LHX5" & sample_info_dg$day == "3")
    sample_info_dg$ko_gene_day4 = as.factor(sample_info_dg$ko_gene == "LHX5" & sample_info_dg$day == "4")
    table(sample_info_dg$ko_gene_day3, sample_info_dg$ko_gene_day4)
    
    dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene_day3 + ko_gene_day4 + run_id + log_q75)   
  
    start.time <- Sys.time()
    dds = DESeq(dds)
    end.time <- Sys.time()

    time.taken <- end.time - start.time
    print(time.taken)
    
    ## association with read-depth
    res_rd = results(dds, name = "log_q75")
    res_rd = as.data.frame(res_rd)
    dim(res_rd)
    res_rd[1:2,]
    
    table(is.na(res_rd$padj))
  
    g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
      geom_histogram(color="darkblue", fill="lightblue", 
                     breaks = seq(0,1,by=0.01)) + 
      ggtitle(paste0(d_group, "\nread depth"))
    print(g0)
  
    ## Rerun the analysis without read-depth if it is not significant for 
    ## most genes, and the number of samples is small
    ## i.e., proportion of non-null in the distribution is smaller than 0.01
    ## (this needs to be restricted to the genes with not NA adjusted pvalue)
    ## or if smaller than 0.1 and the number of samples is not greater than 6

    pi_1_rd = max(0, mean(res_rd$pvalue[!is.na(res_rd$padj)] < 0.05) - 0.05)
    pi_1_rd = max(pi_1_rd, 0, 1 - 2*mean(res_rd$pvalue[!is.na(res_rd$padj)] > 0.5))
    
    print(sprintf("prop of non-null for rd: %.2e", pi_1_rd))
    
    if(pi_1_rd < 0.01 || (ncol(cts_dat_m_dg) <= 6 && pi_1_rd < 0.1 )){
      
      print("rerun DESeq2 without read depth")
      
      include_rd = 0
      
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene_day3 + ko_gene_day4 + run_id) 
      
      dds = DESeq(dds)
      
    }else{
      include_rd = 1
    }

    ## association with run_id
      
    if (length(unique(sample_info_dg$run_id))>1){
      if(include_rd)
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene_day3 + ko_gene_day4 + log_q75)
      else
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene_day3 + ko_gene_day4)
      
      res_lrt = results(dds_lrt)
    
      res_run_id = as.data.frame(res_lrt)
      dim(res_run_id)
      res_run_id[1:2,]
    
      table(is.na(res_run_id$padj))
      
      g0 = ggplot(res_run_id %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, "\nRun ID"))
      print(g0)
    
    }
    
    ## DE analysis for each KO gene
    
    genos = 'LHX5'
    
    for (geno in genos) {
      # add a column to record the number of zeros per test
      test_index = which(sample_info_dg$ko_gene%in%c(geno, "WT"))
      test_ko_gene_vec = sample_info_dg$ko_gene[test_index]
      
      df_size = rbind(df_size, 
                      c(d_group, 
                        paste0(geno, "_", setdiff(sample_info_dg$ko_strategy, 'WT')), 
                        sum(test_ko_gene_vec!="WT"), 
                        sum(test_ko_gene_vec=="WT")))   
    }
    
    for (day in 3:4) {
      for(geno in genos){
        
        res_g1 = results(dds, name = paste0("ko_gene_day", day, "_TRUE_vs_FALSE"))
        res_g1 = signif(data.frame(res_g1), 3)
        
        test_index = which(sample_info_dg$ko_gene%in%c(geno, "WT"))
        cts_dat_m_dg_test = cts_dat_m_dg[, test_index]
        n_zero = rowSums(cts_dat_m_dg_test==0)
        res_g1$n_zeros = n_zero    
        
        # prepare a processed version of output
        res_g1_reduced = res_g1[, c("log2FoldChange", "pvalue", "padj", "n_zeros")]
        res_g1_reduced$padj[which(res_g1_reduced$n_zeros>=4)] = NA
        
        if ((sum(test_ko_gene_vec!="WT")==1)|(sum(test_ko_gene_vec=="WT")==1)){
          res_g1_reduced$padj = NA
        }
        
        res_g1_reduced = res_g1_reduced[, c("log2FoldChange", "pvalue", "padj")]      
    
        n_DE = sum(res_g1_reduced$padj < 0.05 & abs(res_g1_reduced$log2FoldChange) > log2(1.5), 
                  na.rm = TRUE)
        print(paste0(d_group, " under KO gene ", geno, " in day ", day, ":"))
        print(sprintf("number of DE genes: %d", n_DE))

        g1 = ggplot(res_g1_reduced %>% subset(!is.na(padj)), aes(x=pvalue)) + 
          geom_histogram(color="darkblue", fill="lightblue", 
                        breaks = seq(0,1,by=0.01)) + 
          ggtitle(paste0(d_group, "\n", geno))
        print(g1)

        tag1 = sprintf("%s_%s_", geno, setdiff(sample_info_dg$ko_strategy, 'WT'))
        colnames(res_g1) = paste0(tag1, colnames(res_g1))
        colnames(res_g1_reduced) = paste0(tag1, colnames(res_g1_reduced))
        
        
        res_df = data.frame(gene_ID=rownames(res_g1), 
                            res_g1)
        dim(res_df)
        res_df[1:2,]
        
        
        res_reduced_gene_anno = gene_anno[match(rownames(res_g1_reduced), 
                                                gene_anno$ensembl_gene_id), ]
        stopifnot(all(rownames(res_g1_reduced)==res_reduced_gene_anno$ensembl_gene_id))     
        
        # set gene hgnc_symbol that equal "" to NA
        hgnc_symbol_vec = res_reduced_gene_anno$hgnc_symbol
        hgnc_symbol_vec[which(hgnc_symbol_vec=="")] = NA
        
        res_reduced_df = data.frame(gene_ID=rownames(res_g1_reduced), 
                                    hgnc_symbol=hgnc_symbol_vec,
                                    chr=res_reduced_gene_anno$chr, 
                                    strand=res_reduced_gene_anno$strand, 
                                    start=res_reduced_gene_anno$start, 
                                    end=res_reduced_gene_anno$end, 
                                    res_g1_reduced)
        
        print("hgnc_symbol empty string and NA tables:")
        print(table(res_reduced_df$hgnc_symbol=="", useNA="ifany"))
        print(table(is.na(res_reduced_df$hgnc_symbol)))
          
        dim(res_reduced_df)
        res_reduced_df[1:2,]      
        
        setting_name = strsplit(d_group, "_")[[1]]
        setting_name = paste0(setting_name[1], '_day_', day, "_", setting_name[5])
        
        fwrite(res_df, 
              sprintf("%s/%s_%s_%s_DEseq2_raw.tsv", 
                      raw_output_dir, release_name, setting_name, geno), 
              sep="\t")
        fwrite(res_reduced_df, 
              sprintf("%s/%s_%s_%s_DEseq2.tsv", 
                      processed_output_dir, release_name, setting_name, geno), 
              sep="\t")      

      }
    }
  }
  
}

colnames(df_size) = c("DE_group", "knockout_gene_strategy", "n_ko", "n_WT")
df_size = as.data.frame(df_size)

multi_rows = which(table(df_size$knockout_gene_strategy)>1)
multi_rows

df_size[which(df_size$knockout_gene_strategy%in%names(multi_rows)),]

fwrite(df_size, 
       sprintf("%s/%s_DE_n_samples.tsv", output_dir, release_name), 
       sep="\t")
```